<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Human heart failure &mdash; Thor  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/style.css?v=71a31e8f" />
      <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Run Thor on Visium HD data" href="tutorial_runThor_VisiumHD.html" />
    <link rel="prev" title="Ductal carcinoma in situ" href="runThor_DCIS.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            Thor
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Thor</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Thor Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../release_notes.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../FAQ.html">Frequently asked questions</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../API.html">Classes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../API.html#api">API</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="runThor_MOB.html">Mouse olfactory bulb</a></li>
<li class="toctree-l1"><a class="reference internal" href="runThor_DCIS.html">Ductal carcinoma in situ</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Human heart failure</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Import-the-packages">Import the packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Predicting-cell-level-gene-expression-using-Markov-graph-diffusion">Predicting cell-level gene expression using Markov graph diffusion</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Pathway-enrichment-analyses">Pathway enrichment analyses</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Transcription-factor-activity-inference">Transcription factor activity inference</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Semi-supervised-annotation">Semi-supervised annotation</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_runThor_VisiumHD.html">Run Thor on Visium HD data</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_run_cell_communication_commot.html">Cell cell communication</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Mjolnir visualization</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../Mjolnir.html">About Mjolnir</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Mjolnir_installation.html">Mjolnir Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Mjolnir_video_tutorials.html">Mjolnir video tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Mjolnir_test_data.html">Mjolnir test Data</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Thor</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Human heart failure</li>
      <li class="wy-breadcrumbs-aside">
<!--             <a href="../_sources/notebooks/run_thor_analyses_HF.ipynb.txt" rel="nofollow"> View page source</a> -->
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="admonition note">
This page was generated from a Jupyter notebook. You can download the notebook
<a href="javascript:void(0)" onclick="var link = document.createElement('a'); link.download = 'run_thor_analyses_HF.ipynb'; link.href = '../notebooks/run_thor_analyses_HF.ipynb'; link.click(); return false;">here</a> to run it locally in Jupyter.
</div>

<script>
// Alternative download function in case the onclick method doesn't work
function downloadNotebook(e) {
    e.preventDefault();
    var xhr = new XMLHttpRequest();
    xhr.open('GET', '../notebooks/run_thor_analyses_HF.ipynb', true);
    xhr.responseType = 'blob';
    xhr.onload = function() {
        if (this.status === 200) {
            var blob = new Blob([this.response], {type: 'application/x-ipynb+json'});
            var link = document.createElement('a');
            link.href = URL.createObjectURL(blob);
            link.download = 'run_thor_analyses_HF.ipynb';
            link.click();
            setTimeout(function() {
                URL.revokeObjectURL(link.href);
            }, 1500);
        }
    };
    xhr.send();
}

// Add the event listener to all download links
document.addEventListener('DOMContentLoaded', function() {
    var link = document.querySelector('.admonition.note a');
    if (link) {
        link.addEventListener('click', downloadNotebook);
    }
});
</script><section id="Human-heart-failure">
<h1>Human heart failure<a class="headerlink" href="#Human-heart-failure" title="Link to this heading"></a></h1>
<p>In this notebook, we show how to infer cell-level spatial transcriptome and how to perform analyses on the tissue samples from human heart failure patients with myocardial infarction.</p>
<ul class="simple">
<li><p>Thor inference</p></li>
<li><p>Pathway enrichment analysis</p></li>
<li><p>Transcription factor analyses</p></li>
<li><p>Semi-supervised annotation of fibrotic cells (link to external video tutorials to Mjolnir platform, which is to be launched)</p></li>
</ul>
<p>The input data and Thor processed data can be downloaded from our <a class="reference external" href="https://drive.google.com/drive/folders/10k-bDw2zz_3Tdd9OvcSjNNqfKmTNbKqn?usp=sharing">google drive</a> or from the original <a class="reference external" href="https://www.nature.com/articles/s41586-022-05060-x">Nature paper</a>. Aliases (in parentesis) are used in our manuscript of the sample names appeared in the original paper: RZ_GT_P2(RZ1), RZ_P11(RZ2), GT_IZ_P9(IZ1), IZ_P10(IZ2), FZ_GT_P19(FZ1), FZ_P18(FZ2).</p>
<p>For installation of Thor, please refer to <a class="reference internal" href="../installation.html"><span class="doc">this installation guide</span></a>.</p>
<section id="Import-the-packages">
<h2>Import the packages<a class="headerlink" href="#Import-the-packages" title="Link to this heading"></a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">datetime</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">()</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%(name)s</span><span class="s1"> - </span><span class="si">%(levelname)s</span><span class="s1"> - </span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="n">now</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Current Time: </span><span class="si">{</span><span class="n">now</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
root - INFO - Current Time: 2023-12-13 00:13:40.927335
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">config</span> InlineBackend.figure_format = &#39;retina&#39;

<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">matplotlib.colors</span> <span class="kn">import</span> <span class="n">LinearSegmentedColormap</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">decoupler</span> <span class="k">as</span> <span class="nn">dc</span>
<span class="kn">import</span> <span class="nn">warnings</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">scanpy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">150</span><span class="p">,</span> <span class="n">dpi_save</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="s1">&#39;error&#39;</span>

<span class="kn">from</span> <span class="nn">thor.analy</span> <span class="kn">import</span> <span class="n">get_pathway_score</span><span class="p">,</span> <span class="n">get_tf_activity</span>
<span class="kn">from</span> <span class="nn">thor.finest</span> <span class="kn">import</span> <span class="n">fineST</span>
<span class="kn">from</span> <span class="nn">thor.pl</span> <span class="kn">import</span> <span class="n">colors</span>
</pre></div>
</div>
</div>
</section>
<section id="Predicting-cell-level-gene-expression-using-Markov-graph-diffusion">
<h2>Predicting cell-level gene expression using Markov graph diffusion<a class="headerlink" href="#Predicting-cell-level-gene-expression-using-Markov-graph-diffusion" title="Link to this heading"></a></h2>
<p>There are in total of six samples. We use sample “RZ_GT_P2” as an exmaple. All the process and parameters were kept the same.</p>
<p>We also skip the preprocessing of the images and the spatial transcriptome data in this tutorial. Please refer to the other two tutorials for preprocessing. The preprocessed data can be downloaded from the same google drive <a class="reference external" href="https://drive.google.com/drive/folders/1TqDRuqOd6I5RIKN-G-Nu0E2SpORX4BCH?usp=sharing">link</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">samples</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;RZ_GT_P2&#39;</span><span class="p">,</span> <span class="s1">&#39;RZ_P11&#39;</span><span class="p">,</span> <span class="s1">&#39;GT_IZ_P9&#39;</span><span class="p">,</span> <span class="s1">&#39;IZ_P10&#39;</span><span class="p">,</span> <span class="s1">&#39;FZ_GT_P19&#39;</span><span class="p">,</span> <span class="s1">&#39;FZ_P18&#39;</span><span class="p">]</span>

<span class="n">library_names</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;FZ_P20&#39;</span><span class="p">:</span> <span class="s1">&#39;Visium_16_CK294&#39;</span><span class="p">,</span>
    <span class="s1">&#39;FZ_GT_P19&#39;</span><span class="p">:</span> <span class="s1">&#39;Visium_14_CK292&#39;</span><span class="p">,</span>
    <span class="s1">&#39;GT_IZ_P9&#39;</span><span class="p">:</span> <span class="s1">&#39;Visium_9_CK287&#39;</span><span class="p">,</span>
    <span class="s1">&#39;IZ_P10&#39;</span><span class="p">:</span> <span class="s1">&#39;Visium_7_CK285&#39;</span><span class="p">,</span>
    <span class="s1">&#39;RZ_P11&#39;</span><span class="p">:</span> <span class="s1">&#39;Visium_8_CK286&#39;</span><span class="p">,</span>
    <span class="s1">&#39;RZ_GT_P2&#39;</span><span class="p">:</span> <span class="s1">&#39;Visium_5_CK283&#39;</span>
    <span class="p">}</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sn</span> <span class="o">=</span> <span class="n">samples</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>

<span class="c1"># Below are the files you needed to run Thor</span>
<span class="n">image_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;HE_image&quot;</span><span class="p">,</span> <span class="n">sn</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">library_names</span><span class="p">[</span><span class="n">sn</span><span class="p">]</span><span class="si">}</span><span class="s2">.tif&quot;</span><span class="p">)</span>
<span class="n">cell_features_csv_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;WSI_HF&quot;</span><span class="p">,</span> <span class="n">sn</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;features_</span><span class="si">{</span><span class="n">sn</span><span class="si">}</span><span class="s2">.csv&quot;</span><span class="p">)</span>
<span class="n">cell_mask_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;WSI_HF&quot;</span><span class="p">,</span> <span class="n">sn</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;nucleis_</span><span class="si">{</span><span class="n">sn</span><span class="si">}</span><span class="s2">_mask.npz&quot;</span><span class="p">)</span>
<span class="n">spot_adata_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;Spatial_HF&quot;</span><span class="p">,</span> <span class="n">sn</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sn</span><span class="si">}</span><span class="s2">_adata_spots_thor_processed.h5ad&quot;</span><span class="p">)</span>

<span class="n">rz</span> <span class="o">=</span> <span class="n">fineST</span><span class="p">(</span>
    <span class="n">image_path</span><span class="p">,</span>
    <span class="n">sn</span><span class="p">,</span>
    <span class="n">spot_adata_path</span><span class="o">=</span><span class="n">spot_adata_path</span><span class="p">,</span>
    <span class="n">cell_features_csv_path</span><span class="o">=</span><span class="n">cell_features_csv_path</span><span class="p">,</span>
    <span class="n">recipe</span><span class="o">=</span><span class="s1">&#39;gene&#39;</span>

    <span class="p">)</span>
<span class="n">rz</span><span class="o">.</span><span class="n">prepare_input</span><span class="p">(</span><span class="n">mapping_margin</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-cyan-fg">[12/13/23 00:13:46]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Please check alignment of cells and spots
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: The first two columns in the node_feat DataFrame need to match the spatial coordinates from obsm<span class="ansi-bold">[</span><span class="ansi-green-fg">&#39;spatial&#39;</span><span class="ansi-bold">]</span>.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Mapping cells to the closest spots within <span class="ansi-cyan-intense-fg ansi-bold">10</span> x the spot radius
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_run_thor_analyses_HF_7_1.png" class="no-scaled-link" src="../_images/notebooks_run_thor_analyses_HF_7_1.png" style="width: 926px; height: 854px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rz</span><span class="o">.</span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 32994 × 14467
    obs: &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;percent.mt&#39;, &#39;Adipocyte&#39;, &#39;Cardiomyocyte&#39;, &#39;Endothelial&#39;, &#39;Fibroblast&#39;, &#39;Lymphoid&#39;, &#39;Mast&#39;, &#39;Myeloid&#39;, &#39;Neuronal&#39;, &#39;Pericyte&#39;, &#39;Cycling.cells&#39;, &#39;vSMCs&#39;, &#39;cell_type_original&#39;, &#39;assay_ontology_term_id&#39;, &#39;cell_type_ontology_term_id&#39;, &#39;development_stage_ontology_term_id&#39;, &#39;disease_ontology_term_id&#39;, &#39;ethnicity_ontology_term_id&#39;, &#39;is_primary_data&#39;, &#39;organism_ontology_term_id&#39;, &#39;sex_ontology_term_id&#39;, &#39;tissue_ontology_term_id&#39;, &#39;spot_barcodes&#39;, &#39;x&#39;, &#39;y&#39;, &#39;mean_gray&#39;, &#39;std_gray&#39;, &#39;entropy_img&#39;, &#39;mean_r&#39;, &#39;mean_g&#39;, &#39;mean_b&#39;, &#39;std_r&#39;, &#39;std_g&#39;, &#39;std_b&#39;, &#39;spot_heterogeneity&#39;
    var: &#39;features&#39;
    uns: &#39;X_approximate_distribution&#39;, &#39;X_normalization&#39;, &#39;default_embedding&#39;, &#39;schema_version&#39;, &#39;spatial&#39;, &#39;title&#39;, &#39;cell_image_props&#39;
    obsm: &#39;X_pca&#39;, &#39;X_spatial&#39;, &#39;X_umap&#39;, &#39;spatial&#39;
</pre></div></div>
</div>
<p>Run time ~ 10 mins for 32994 cells and 14467 genes on a MacbookPro M2 laptop.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rz</span><span class="o">.</span><span class="n">set_genes_for_prediction</span><span class="p">(</span><span class="n">genes_selection_key</span><span class="o">=</span><span class="s1">&#39;all&#39;</span><span class="p">)</span>

<span class="n">rz</span><span class="o">.</span><span class="n">set_params</span><span class="p">(</span>
    <span class="n">out_prefix</span><span class="o">=</span><span class="s2">&quot;fineST&quot;</span><span class="p">,</span>
    <span class="n">is_rawCounts</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">layer</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">is_rawCount</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">write_freq</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
    <span class="n">n_iter</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
    <span class="n">smoothing_scale</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
    <span class="n">node_features_obs_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;spot_heterogeneity&#39;</span><span class="p">],</span>
    <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">geom_morph_ratio</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">inflation_percentage</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">regulate_expression_mean</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">stochastic_expression_neighbors_level</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">n_jobs</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>

<span class="n">rz</span><span class="o">.</span><span class="n">predict_gene_expression</span><span class="p">()</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-cyan-fg">[12/12/23 23:00:47]</span><span class="ansi-cyan-fg"> </span><span class="ansi-red-fg">WARNING </span> Thor: Using all the genes. This may not be optimal and can be slow.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Using mode gene
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Construct SNN with morphological features: <span class="ansi-bold">[</span><span class="ansi-green-fg">&#39;mean_gray&#39;</span>, <span class="ansi-green-fg">&#39;std_gray&#39;</span>, <span class="ansi-green-fg">&#39;entropy_img&#39;</span>, <span class="ansi-green-fg">&#39;mean_r&#39;</span>, <span class="ansi-green-fg">&#39;mean_g&#39;</span>, <span class="ansi-green-fg">&#39;mean_b&#39;</span>,
<span class="ansi-cyan-fg">                    </span>         <span class="ansi-green-fg">&#39;std_r&#39;</span>, <span class="ansi-green-fg">&#39;std_g&#39;</span>, <span class="ansi-green-fg">&#39;std_b&#39;</span><span class="ansi-bold">]</span>.
<span class="ansi-cyan-fg">[12/12/23 23:00:48]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Finish constructing SNN
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Add adata.obsp<span class="ansi-bold">[</span><span class="ansi-green-fg">&#34;snn_connectivities&#34;</span><span class="ansi-bold">]</span>
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Add adata.obsp<span class="ansi-bold">[</span><span class="ansi-green-fg">&#34;snn_knn_connectivities&#34;</span><span class="ansi-bold">]</span>
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Add adata.uns<span class="ansi-bold">[</span><span class="ansi-green-fg">&#34;snn&#34;</span><span class="ansi-bold">]</span>
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Weigh cells according to the spot heterogeneity.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Promote flow of information from more homogeneous cells to less.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Eliminate low quality edges <span class="ansi-bold">(</span>&lt;<span class="ansi-cyan-intense-fg ansi-bold">0.1</span><span class="ansi-bold">)</span> between cells.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Compute transition matrix
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: self weight scale is set to: <span class="ansi-cyan-intense-fg ansi-bold">0.200</span>
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Added transition matrix to adata.obsp<span class="ansi-bold">[</span><span class="ansi-green-fg">&#34;snn_transition_matrix&#34;</span><span class="ansi-bold">]</span>
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;spot_barcodes&#39; as categorical
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-cyan-fg">[12/12/23 23:00:49]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: fineST estimation starts.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Creating piles: 100%|<span class="ansi-green-fg">██████████</span>| 8/8 [00:20&lt;00:00,  2.60s/it]
MCMC Iteration: 100%|██████████| 20/20 [07:06&lt;00:00, 21.31s/it]
MCMC Iteration: 100%|██████████| 20/20 [07:09&lt;00:00, 21.46s/it]
MCMC Iteration: 100%|██████████| 20/20 [07:08&lt;00:00, 21.43s/it]
MCMC Iteration: 100%|██████████| 20/20 [07:04&lt;00:00, 21.22s/it]
MCMC Iteration: 100%|██████████| 20/20 [07:12&lt;00:00, 21.63s/it]
MCMC Iteration: 100%|██████████| 20/20 [07:10&lt;00:00, 21.50s/it]
MCMC Iteration: 100%|██████████| 20/20 [07:03&lt;00:00, 21.20s/it]
MCMC Iteration: 100%|██████████| 20/20 [07:01&lt;00:00, 21.08s/it]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-cyan-fg">[12/12/23 23:10:33]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: fineST estimation finished.
<span class="ansi-cyan-fg">                    </span>
<span class="ansi-cyan-fg">                    </span>
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad_thor</span> <span class="o">=</span> <span class="n">rz</span><span class="o">.</span><span class="n">load_result</span><span class="p">(</span><span class="s2">&quot;fineST_20.npz&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">ad_thor</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">ad_thor</span><span class="p">)</span>
<span class="n">ad_thor</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 32994 × 14467
    obs: &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;percent.mt&#39;, &#39;Adipocyte&#39;, &#39;Cardiomyocyte&#39;, &#39;Endothelial&#39;, &#39;Fibroblast&#39;, &#39;Lymphoid&#39;, &#39;Mast&#39;, &#39;Myeloid&#39;, &#39;Neuronal&#39;, &#39;Pericyte&#39;, &#39;Cycling.cells&#39;, &#39;vSMCs&#39;, &#39;cell_type_original&#39;, &#39;assay_ontology_term_id&#39;, &#39;cell_type_ontology_term_id&#39;, &#39;development_stage_ontology_term_id&#39;, &#39;disease_ontology_term_id&#39;, &#39;ethnicity_ontology_term_id&#39;, &#39;is_primary_data&#39;, &#39;organism_ontology_term_id&#39;, &#39;sex_ontology_term_id&#39;, &#39;tissue_ontology_term_id&#39;, &#39;spot_barcodes&#39;, &#39;x&#39;, &#39;y&#39;, &#39;mean_gray&#39;, &#39;std_gray&#39;, &#39;entropy_img&#39;, &#39;mean_r&#39;, &#39;mean_g&#39;, &#39;mean_b&#39;, &#39;std_r&#39;, &#39;std_g&#39;, &#39;std_b&#39;, &#39;spot_heterogeneity&#39;, &#39;node_weights&#39;
    var: &#39;features&#39;, &#39;used_for_prediction&#39;, &#39;used_for_reduced&#39;, &#39;used_for_vae&#39;
    uns: &#39;X_approximate_distribution&#39;, &#39;X_normalization&#39;, &#39;default_embedding&#39;, &#39;schema_version&#39;, &#39;spatial&#39;, &#39;title&#39;, &#39;cell_image_props&#39;, &#39;snn&#39;, &#39;log1p&#39;
    obsm: &#39;X_pca&#39;, &#39;X_spatial&#39;, &#39;X_umap&#39;, &#39;spatial&#39;
    obsp: &#39;snn_connectivities&#39;, &#39;snn_knn_connectivities&#39;, &#39;snn_transition_matrix&#39;
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">palette</span> <span class="o">=</span> <span class="n">colors</span><span class="o">.</span><span class="n">continuous_palettes</span><span class="p">[</span><span class="s1">&#39;blueYellow&#39;</span><span class="p">]</span>
<span class="n">cmap</span> <span class="o">=</span> <span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span><span class="s1">&#39;my_cmap&#39;</span><span class="p">,</span> <span class="n">palette</span><span class="p">,</span> <span class="n">N</span><span class="o">=</span><span class="mi">256</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">ad_thor</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;PDGFRA&#39;</span><span class="p">,</span> <span class="s1">&#39;CASQ2&#39;</span><span class="p">,</span> <span class="s1">&#39;FBLN2&#39;</span><span class="p">,</span> <span class="s1">&#39;PKP2&#39;</span><span class="p">],</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="s1">&#39;p1&#39;</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="s1">&#39;p99&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_run_thor_analyses_HF_13_0.png" class="no-scaled-link" src="../_images/notebooks_run_thor_analyses_HF_13_0.png" style="width: 1200px; height: 280px;" />
</div>
</div>
</section>
<section id="Pathway-enrichment-analyses">
<h2>Pathway enrichment analyses<a class="headerlink" href="#Pathway-enrichment-analyses" title="Link to this heading"></a></h2>
<p>Get the gene sets for the GO_BP pathways</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">decoupler</span> <span class="k">as</span> <span class="nn">dc</span>

<span class="n">msigdb</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">get_resource</span><span class="p">(</span><span class="s1">&#39;MSigDB&#39;</span><span class="p">)</span>
<span class="n">msigdb</span> <span class="o">=</span> <span class="n">msigdb</span><span class="p">[</span><span class="n">msigdb</span><span class="p">[</span><span class="s1">&#39;collection&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;go_biological_process&#39;</span><span class="p">]</span>
<span class="n">msigdb</span> <span class="o">=</span> <span class="n">msigdb</span><span class="p">[</span><span class="o">~</span><span class="n">msigdb</span><span class="o">.</span><span class="n">duplicated</span><span class="p">([</span><span class="s1">&#39;geneset&#39;</span><span class="p">,</span> <span class="s1">&#39;genesymbol&#39;</span><span class="p">])]</span>
<span class="n">msigdb</span><span class="p">[</span><span class="s1">&#39;geneset&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;GOBP_&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">msigdb</span><span class="p">[</span><span class="s1">&#39;geneset&#39;</span><span class="p">]]</span>
<span class="n">msigdb</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
root - INFO - Downloading data from `https://omnipathdb.org/queries/enzsub?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/queries/interactions?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/queries/complexes?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/queries/annotations?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/queries/intercell?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/about?format=text`
root - INFO - Downloading annotations for all proteins from the following resources: `[&#39;MSigDB&#39;]`
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>genesymbol</th>
      <th>collection</th>
      <th>geneset</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>33</th>
      <td>MAFF</td>
      <td>go_biological_process</td>
      <td>EMBRYO_DEVELOPMENT</td>
    </tr>
    <tr>
      <th>44</th>
      <td>MAFF</td>
      <td>go_biological_process</td>
      <td>POSITIVE_REGULATION_OF_RNA_METABOLIC_PROCESS</td>
    </tr>
    <tr>
      <th>82</th>
      <td>MAFF</td>
      <td>go_biological_process</td>
      <td>REGULATION_OF_EPITHELIAL_CELL_DIFFERENTIATION</td>
    </tr>
    <tr>
      <th>94</th>
      <td>MAFF</td>
      <td>go_biological_process</td>
      <td>EMBRYO_DEVELOPMENT_ENDING_IN_BIRTH_OR_EGG_HATC...</td>
    </tr>
    <tr>
      <th>108</th>
      <td>MAFF</td>
      <td>go_biological_process</td>
      <td>IN_UTERO_EMBRYONIC_DEVELOPMENT</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>3838543</th>
      <td>PRAMEF22</td>
      <td>go_biological_process</td>
      <td>POSITIVE_REGULATION_OF_CELL_POPULATION_PROLIFE...</td>
    </tr>
    <tr>
      <th>3838544</th>
      <td>PRAMEF22</td>
      <td>go_biological_process</td>
      <td>APOPTOTIC_PROCESS</td>
    </tr>
    <tr>
      <th>3838545</th>
      <td>PRAMEF22</td>
      <td>go_biological_process</td>
      <td>REGULATION_OF_CELL_DEATH</td>
    </tr>
    <tr>
      <th>3838546</th>
      <td>PRAMEF22</td>
      <td>go_biological_process</td>
      <td>NEGATIVE_REGULATION_OF_DEVELOPMENTAL_PROCESS</td>
    </tr>
    <tr>
      <th>3838547</th>
      <td>PRAMEF22</td>
      <td>go_biological_process</td>
      <td>NEGATIVE_REGULATION_OF_CELL_DEATH</td>
    </tr>
  </tbody>
</table>
<p>620627 rows × 3 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ptw_list</span> <span class="o">=</span> <span class="p">[</span>
    <span class="s1">&#39;POSITIVE_REGULATION_OF_T_CELL_PROLIFERATION&#39;</span><span class="p">,</span>
    <span class="s1">&#39;POSITIVE_REGULATION_OF_INFLAMMATORY_RESPONSE&#39;</span><span class="p">,</span>
    <span class="s1">&#39;POSITIVE_REGULATION_OF_IMMUNE_EFFECTOR_PROCESS&#39;</span><span class="p">,</span>
    <span class="s1">&#39;POSITIVE_REGULATION_OF_B_CELL_ACTIVATION&#39;</span><span class="p">,</span>
    <span class="s1">&#39;POSITIVE_REGULATION_OF_FIBROBLAST_PROLIFERATION&#39;</span><span class="p">,</span>
    <span class="s1">&#39;COLLAGEN_FIBRIL_ORGANIZATION&#39;</span><span class="p">,</span>
    <span class="s1">&#39;CARDIAC_MUSCLE_CONTRACTION&#39;</span>
    <span class="p">]</span>
<span class="n">msigdb</span> <span class="o">=</span> <span class="n">msigdb</span><span class="p">[</span><span class="n">msigdb</span><span class="o">.</span><span class="n">geneset</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">ptw_list</span><span class="p">)]</span>
<span class="n">msigdb</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>genesymbol</th>
      <th>collection</th>
      <th>geneset</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>3181</th>
      <td>GPR183</td>
      <td>go_biological_process</td>
      <td>POSITIVE_REGULATION_OF_B_CELL_ACTIVATION</td>
    </tr>
    <tr>
      <th>5388</th>
      <td>RIPK2</td>
      <td>go_biological_process</td>
      <td>POSITIVE_REGULATION_OF_IMMUNE_EFFECTOR_PROCESS</td>
    </tr>
    <tr>
      <th>5673</th>
      <td>RIPK2</td>
      <td>go_biological_process</td>
      <td>POSITIVE_REGULATION_OF_T_CELL_PROLIFERATION</td>
    </tr>
    <tr>
      <th>12886</th>
      <td>CEBPB</td>
      <td>go_biological_process</td>
      <td>POSITIVE_REGULATION_OF_INFLAMMATORY_RESPONSE</td>
    </tr>
    <tr>
      <th>14576</th>
      <td>SPHK1</td>
      <td>go_biological_process</td>
      <td>POSITIVE_REGULATION_OF_FIBROBLAST_PROLIFERATION</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>3815299</th>
      <td>SASH3</td>
      <td>go_biological_process</td>
      <td>POSITIVE_REGULATION_OF_IMMUNE_EFFECTOR_PROCESS</td>
    </tr>
    <tr>
      <th>3815331</th>
      <td>SASH3</td>
      <td>go_biological_process</td>
      <td>POSITIVE_REGULATION_OF_T_CELL_PROLIFERATION</td>
    </tr>
    <tr>
      <th>3815499</th>
      <td>SASH3</td>
      <td>go_biological_process</td>
      <td>POSITIVE_REGULATION_OF_B_CELL_ACTIVATION</td>
    </tr>
    <tr>
      <th>3826247</th>
      <td>TAFAZZIN</td>
      <td>go_biological_process</td>
      <td>CARDIAC_MUSCLE_CONTRACTION</td>
    </tr>
    <tr>
      <th>3838037</th>
      <td>HLA-DRB3</td>
      <td>go_biological_process</td>
      <td>POSITIVE_REGULATION_OF_IMMUNE_EFFECTOR_PROCESS</td>
    </tr>
  </tbody>
</table>
<p>843 rows × 3 columns</p>
</div></div>
</div>
<p>Calculate the enrichment score using decoupler</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">acts</span> <span class="o">=</span> <span class="n">get_pathway_score</span><span class="p">(</span><span class="n">ad_thor</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">net_df</span><span class="o">=</span><span class="n">msigdb</span><span class="p">)</span>
<span class="n">acts</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Running ora on mat with 32994 samples and 14467 targets for 7 sources.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 32994/32994 [00:28&lt;00:00, 1175.86it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 32994 × 7
    obs: &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;percent.mt&#39;, &#39;Adipocyte&#39;, &#39;Cardiomyocyte&#39;, &#39;Endothelial&#39;, &#39;Fibroblast&#39;, &#39;Lymphoid&#39;, &#39;Mast&#39;, &#39;Myeloid&#39;, &#39;Neuronal&#39;, &#39;Pericyte&#39;, &#39;Cycling.cells&#39;, &#39;vSMCs&#39;, &#39;cell_type_original&#39;, &#39;assay_ontology_term_id&#39;, &#39;cell_type_ontology_term_id&#39;, &#39;development_stage_ontology_term_id&#39;, &#39;disease_ontology_term_id&#39;, &#39;ethnicity_ontology_term_id&#39;, &#39;is_primary_data&#39;, &#39;organism_ontology_term_id&#39;, &#39;sex_ontology_term_id&#39;, &#39;tissue_ontology_term_id&#39;, &#39;spot_barcodes&#39;, &#39;x&#39;, &#39;y&#39;, &#39;mean_gray&#39;, &#39;std_gray&#39;, &#39;entropy_img&#39;, &#39;mean_r&#39;, &#39;mean_g&#39;, &#39;mean_b&#39;, &#39;std_r&#39;, &#39;std_g&#39;, &#39;std_b&#39;, &#39;spot_heterogeneity&#39;, &#39;node_weights&#39;
    uns: &#39;X_approximate_distribution&#39;, &#39;X_normalization&#39;, &#39;cell_image_props&#39;, &#39;default_embedding&#39;, &#39;log1p&#39;, &#39;schema_version&#39;, &#39;snn&#39;, &#39;spatial&#39;, &#39;title&#39;
    obsm: &#39;X_pca&#39;, &#39;X_spatial&#39;, &#39;X_umap&#39;, &#39;ora_estimate&#39;, &#39;ora_pvals&#39;, &#39;spatial&#39;, &#39;msigdb_ora_estimate&#39;, &#39;msigdb_ora_pvals&#39;
    layers: &#39;smoothed&#39;
</pre></div></div>
</div>
<p>The pathway enrichment scores can be visualized on our interactive browser Mjolnir or using SCANPY as follows</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">pathway</span> <span class="ow">in</span> <span class="n">ptw_list</span><span class="p">:</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">acts</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">pathway</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s1">&#39;smoothed&#39;</span><span class="p">,</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;rainbow&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">pathway</span><span class="o">.</span><span class="n">lower</span><span class="p">(),</span> <span class="n">colorbar_loc</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">alpha_img</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_run_thor_analyses_HF_21_0.png" class="no-scaled-link" src="../_images/notebooks_run_thor_analyses_HF_21_0.png" style="width: 566px; height: 519px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_run_thor_analyses_HF_21_1.png" class="no-scaled-link" src="../_images/notebooks_run_thor_analyses_HF_21_1.png" style="width: 633px; height: 519px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_run_thor_analyses_HF_21_2.png" class="no-scaled-link" src="../_images/notebooks_run_thor_analyses_HF_21_2.png" style="width: 665px; height: 519px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_run_thor_analyses_HF_21_3.png" class="no-scaled-link" src="../_images/notebooks_run_thor_analyses_HF_21_3.png" style="width: 545px; height: 519px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_run_thor_analyses_HF_21_4.png" class="no-scaled-link" src="../_images/notebooks_run_thor_analyses_HF_21_4.png" style="width: 617px; height: 519px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_run_thor_analyses_HF_21_5.png" class="no-scaled-link" src="../_images/notebooks_run_thor_analyses_HF_21_5.png" style="width: 469px; height: 519px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_run_thor_analyses_HF_21_6.png" class="no-scaled-link" src="../_images/notebooks_run_thor_analyses_HF_21_6.png" style="width: 469px; height: 519px;" />
</div>
</div>
</section>
<section id="Transcription-factor-activity-inference">
<h2>Transcription factor activity inference<a class="headerlink" href="#Transcription-factor-activity-inference" title="Link to this heading"></a></h2>
<p>Download the CollecTRI database.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">net</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">get_collectri</span><span class="p">(</span><span class="n">organism</span><span class="o">=</span><span class="s1">&#39;human&#39;</span><span class="p">,</span> <span class="n">split_complexes</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">net</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
root - INFO - Downloading data from `https://omnipathdb.org/queries/enzsub?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/queries/interactions?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/queries/complexes?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/queries/annotations?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/queries/intercell?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/about?format=text`
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>source</th>
      <th>target</th>
      <th>weight</th>
      <th>PMID</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>MYC</td>
      <td>TERT</td>
      <td>1</td>
      <td>10022128;10491298;10606235;10637317;10723141;1...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>SPI1</td>
      <td>BGLAP</td>
      <td>1</td>
      <td>10022617</td>
    </tr>
    <tr>
      <th>2</th>
      <td>SMAD3</td>
      <td>JUN</td>
      <td>1</td>
      <td>10022869;12374795</td>
    </tr>
    <tr>
      <th>3</th>
      <td>SMAD4</td>
      <td>JUN</td>
      <td>1</td>
      <td>10022869;12374795</td>
    </tr>
    <tr>
      <th>4</th>
      <td>STAT5A</td>
      <td>IL2</td>
      <td>1</td>
      <td>10022878;11435608;17182565;17911616;22854263;2...</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>43173</th>
      <td>NFKB</td>
      <td>hsa-miR-143-3p</td>
      <td>1</td>
      <td>19472311</td>
    </tr>
    <tr>
      <th>43174</th>
      <td>AP1</td>
      <td>hsa-miR-206</td>
      <td>1</td>
      <td>19721712</td>
    </tr>
    <tr>
      <th>43175</th>
      <td>NFKB</td>
      <td>hsa-miR-21-5p</td>
      <td>1</td>
      <td>20813833;22387281</td>
    </tr>
    <tr>
      <th>43176</th>
      <td>NFKB</td>
      <td>hsa-miR-224-5p</td>
      <td>1</td>
      <td>23474441;23988648</td>
    </tr>
    <tr>
      <th>43177</th>
      <td>AP1</td>
      <td>hsa-miR-144</td>
      <td>1</td>
      <td>23546882</td>
    </tr>
  </tbody>
</table>
<p>43178 rows × 4 columns</p>
</div></div>
</div>
<p>Calculate the TF activity based on regulated gene expression using decoupler</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">acts</span> <span class="o">=</span> <span class="n">get_tf_activity</span><span class="p">(</span><span class="n">ad_thor</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">net_df</span><span class="o">=</span><span class="n">net</span><span class="p">)</span>
<span class="n">acts</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Running ulm on mat with 32994 samples and 14467 targets for 659 sources.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 4/4 [00:03&lt;00:00,  1.20it/s]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 32994 × 659
    obs: &#39;n_counts&#39;, &#39;n_genes&#39;, &#39;percent.mt&#39;, &#39;Adipocyte&#39;, &#39;Cardiomyocyte&#39;, &#39;Endothelial&#39;, &#39;Fibroblast&#39;, &#39;Lymphoid&#39;, &#39;Mast&#39;, &#39;Myeloid&#39;, &#39;Neuronal&#39;, &#39;Pericyte&#39;, &#39;Cycling.cells&#39;, &#39;vSMCs&#39;, &#39;cell_type_original&#39;, &#39;assay_ontology_term_id&#39;, &#39;cell_type_ontology_term_id&#39;, &#39;development_stage_ontology_term_id&#39;, &#39;disease_ontology_term_id&#39;, &#39;ethnicity_ontology_term_id&#39;, &#39;is_primary_data&#39;, &#39;organism_ontology_term_id&#39;, &#39;sex_ontology_term_id&#39;, &#39;tissue_ontology_term_id&#39;, &#39;spot_barcodes&#39;, &#39;x&#39;, &#39;y&#39;, &#39;mean_gray&#39;, &#39;std_gray&#39;, &#39;entropy_img&#39;, &#39;mean_r&#39;, &#39;mean_g&#39;, &#39;mean_b&#39;, &#39;std_r&#39;, &#39;std_g&#39;, &#39;std_b&#39;, &#39;spot_heterogeneity&#39;, &#39;node_weights&#39;
    uns: &#39;X_approximate_distribution&#39;, &#39;X_normalization&#39;, &#39;cell_image_props&#39;, &#39;default_embedding&#39;, &#39;log1p&#39;, &#39;schema_version&#39;, &#39;snn&#39;, &#39;spatial&#39;, &#39;title&#39;
    obsm: &#39;X_pca&#39;, &#39;X_spatial&#39;, &#39;X_umap&#39;, &#39;ora_estimate&#39;, &#39;ora_pvals&#39;, &#39;spatial&#39;, &#39;ulm_estimate&#39;, &#39;ulm_pvals&#39;, &#39;msigdb_ulm_estimate&#39;, &#39;msigdb_ulm_pvals&#39;
    layers: &#39;smoothed&#39;
</pre></div></div>
</div>
<p>The TF activity score can be visualized on our interactive browser Mjolnir or using SCANPY as follows</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tf_list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;MYB&#39;</span><span class="p">,</span> <span class="s1">&#39;STAT4&#39;</span><span class="p">,</span> <span class="s1">&#39;SMAD3&#39;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">tf</span> <span class="ow">in</span> <span class="n">tf_list</span><span class="p">:</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">acts</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">tf</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="s1">&#39;smoothed&#39;</span><span class="p">,</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;rainbow&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="n">tf</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">alpha_img</span><span class="o">=</span><span class="mf">0.7</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_run_thor_analyses_HF_28_0.png" class="no-scaled-link" src="../_images/notebooks_run_thor_analyses_HF_28_0.png" style="width: 506px; height: 512px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_run_thor_analyses_HF_28_1.png" class="no-scaled-link" src="../_images/notebooks_run_thor_analyses_HF_28_1.png" style="width: 523px; height: 521px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_run_thor_analyses_HF_28_2.png" class="no-scaled-link" src="../_images/notebooks_run_thor_analyses_HF_28_2.png" style="width: 522px; height: 523px;" />
</div>
</div>
</section>
<section id="Semi-supervised-annotation">
<h2>Semi-supervised annotation<a class="headerlink" href="#Semi-supervised-annotation" title="Link to this heading"></a></h2>
<p>The semi-supervised annotation of similar cells can be performed on our web-based platform Mjolnir. Below are three video tips for how to use mjolnir.</p>
<ul class="simple">
<li><p><a class="reference external" href="https://drive.google.com/file/d/1g_zGoHohbeXP2RFN_AVAhTQqSAKY5qcj/view?usp=sharing">Visualization of gene expression at spot and cell level</a></p></li>
<li><p><a class="reference external" href="https://drive.google.com/file/d/1e04zKErfZnXknPB3U7eENIfyq_o-RaBS/view?usp=sharing">Visualization of pathway enrichment heatmaps</a></p></li>
<li><p><a class="reference external" href="https://drive.google.com/file/d/1gZYKZTNUnzS_qd2YZrGEoWd7RYhoIAjd/view?usp=sharing">Semi-supervised annotation of cells</a></p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="runThor_DCIS.html" class="btn btn-neutral float-left" title="Ductal carcinoma in situ" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="tutorial_runThor_VisiumHD.html" class="btn btn-neutral float-right" title="Run Thor on Visium HD data" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Wang Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>