<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Ductal carcinoma in situ &mdash; Thor  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/style.css?v=71a31e8f" />
      <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Human heart failure" href="run_thor_analyses_HF.html" />
    <link rel="prev" title="Mouse olfactory bulb" href="runThor_MOB.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            Thor
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Thor</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Thor Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../release_notes.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../FAQ.html">Frequently asked questions</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../API.html">Classes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../API.html#api">API</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="runThor_MOB.html">Mouse olfactory bulb</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Ductal carcinoma in situ</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Run-Thor">Run Thor</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Import-the-packages">Import the packages</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Preprocessing">Preprocessing</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Predicting-cell-level-gene-expression-using-Thor-diffusion">Predicting cell-level gene expression using Thor diffusion</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Advance-analysis">Advance analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Import-packages">Import packages</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Load-the-image-(for-visualization)-and-Thor-predicted-cell-level-transcriptome-result">Load the image (for visualization) and Thor-predicted cell-level transcriptome result</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Visualize-genes-at-tissue-and-cell-level">Visualize genes at tissue and cell level</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Expression-profile-of-oncogene-and-tumor-suppressor-gene">Expression profile of oncogene and tumor suppressor gene</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Hallmark-pathway-enrichment">Hallmark pathway enrichment</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Copy-number-variation">Copy number variation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Tertiary-Lymphoid-Structures(TLS)-quantification">Tertiary Lymphoid Structures(TLS) quantification</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="run_thor_analyses_HF.html">Human heart failure</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_runThor_VisiumHD.html">Run Thor on Visium HD data</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_run_cell_communication_commot.html">Cell cell communication</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Mjolnir visualization</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../Mjolnir.html">About Mjolnir</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Mjolnir_installation.html">Mjolnir Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Mjolnir_video_tutorials.html">Mjolnir video tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Mjolnir_test_data.html">Mjolnir test Data</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Thor</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Ductal carcinoma in situ</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/notebooks/runThor_DCIS.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="admonition note">
This page was generated from a Jupyter notebook. You can download the notebook
<a href="javascript:void(0)" onclick="var link = document.createElement('a'); link.download = 'runThor_DCIS.ipynb'; link.href = '../notebooks/runThor_DCIS.ipynb'; link.click(); return false;">here</a> to run it locally in Jupyter.
</div>

<script>
// Alternative download function in case the onclick method doesn't work
function downloadNotebook(e) {
    e.preventDefault();
    var xhr = new XMLHttpRequest();
    xhr.open('GET', '../notebooks/runThor_DCIS.ipynb', true);
    xhr.responseType = 'blob';
    xhr.onload = function() {
        if (this.status === 200) {
            var blob = new Blob([this.response], {type: 'application/x-ipynb+json'});
            var link = document.createElement('a');
            link.href = URL.createObjectURL(blob);
            link.download = 'runThor_DCIS.ipynb';
            link.click();
            setTimeout(function() {
                URL.revokeObjectURL(link.href);
            }, 1500);
        }
    };
    xhr.send();
}

// Add the event listener to all download links
document.addEventListener('DOMContentLoaded', function() {
    var link = document.querySelector('.admonition.note a');
    if (link) {
        link.addEventListener('click', downloadNotebook);
    }
});
</script><section id="Ductal-carcinoma-in-situ">
<h1>Ductal carcinoma in situ<a class="headerlink" href="#Ductal-carcinoma-in-situ" title="Link to this heading"></a></h1>
<p>Thor performs unbiased screening of breast cancer hallmarks in Ductal Carcinoma in situ sample and reveals heterogeneity of immune response.</p>
<section id="Run-Thor">
<h2>Run Thor<a class="headerlink" href="#Run-Thor" title="Link to this heading"></a></h2>
<p>In this tutorial, we show inference of cell-level spatial gene expression from the <a class="reference external" href="https://www.10xgenomics.com/spatial-transcriptomics/">Visium</a> spot-level spatial transcriptome and a whole slide image (H&amp;E staining) using Thor.</p>
<p>The spatial dataset is Human Breast Cancer: Ductal Carcinoma In Situ, Invasive Carcinoma (FFPE). The input data are downloaded from 10x Genomics <a class="reference external" href="https://www.10xgenomics.com/resources/datasets/human-breast-cancer-ductal-carcinoma-in-situ-invasive-carcinoma-ffpe-1-standard-1-3-0">website</a>.</p>
<p>Thor processed data can be downloaded directly from <a class="reference external" href="https://drive.google.com/drive/folders/18Wu1k09nQQ7gK40oRL5AWHDy-bDBc9en?usp=sharing">google drive</a>.</p>
<p>For installation of Thor, please refer to <a class="reference internal" href="../installation.html"><span class="doc">this installation guide</span></a>.</p>
<section id="Import-the-packages">
<h3>Import the packages<a class="headerlink" href="#Import-the-packages" title="Link to this heading"></a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">datetime</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">()</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%(name)s</span><span class="s1"> - </span><span class="si">%(levelname)s</span><span class="s1"> - </span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="n">now</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Current Time: </span><span class="si">{</span><span class="n">now</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
root - INFO - Current Time: 2023-12-13 16:53:16.118962
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">config</span> InlineBackend.figure_format = &#39;retina&#39;


<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">scanpy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">dpi_save</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="s1">&#39;error&#39;</span>


<span class="kn">from</span> <span class="nn">thor.pp</span> <span class="kn">import</span> <span class="n">WholeSlideImage</span><span class="p">,</span> <span class="n">Spatial</span>
<span class="kn">from</span> <span class="nn">thor.finest</span> <span class="kn">import</span> <span class="n">fineST</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
numexpr.utils - INFO - Note: NumExpr detected 12 cores but &#34;NUMEXPR_MAX_THREADS&#34; not set, so enforcing safe limit of 8.
numexpr.utils - INFO - NumExpr defaulting to 8 threads.
</pre></div></div>
</div>
</section>
<section id="Preprocessing">
<h3>Preprocessing<a class="headerlink" href="#Preprocessing" title="Link to this heading"></a></h3>
<p>Preprocessing the whole slide image, including cell segmentation and morphological feature extraction. If not specified, the segmentation data will be saved in a directory (created by Thor) WSI_{sn}.</p>
<p>The WSI can be downloaded directly from <a class="reference external" href="https://drive.google.com/file/d/1ov3ivbnAjg-PTElAYMWdgt8DLX0AJ6gE/view?usp=sharing">google drive</a>.</p>
<p>The outcomes are: cell mask and the cell feature.csv</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sn</span> <span class="o">=</span> <span class="s1">&#39;DCIS_10x&#39;</span>
<span class="n">image_path</span> <span class="o">=</span> <span class="s2">&quot;Visium_FFPE_Human_Breast_Cancer_image.tif&quot;</span>

<span class="n">wsi</span> <span class="o">=</span> <span class="n">WholeSlideImage</span><span class="p">(</span><span class="n">image_path</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">sn</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Here we use Cellpose to segment the nuclei. Notice the segmentation can take a long time without GPU. The output file is a <code class="docutils literal notranslate"><span class="pre">n_pixel_row</span></code> x <code class="docutils literal notranslate"><span class="pre">n_pixel_col</span></code> matrix. We’ll skip this step and load the pre-computed segmentation results from <a class="reference external" href="https://drive.google.com/drive/folders/1ujxAGOWtDexImGGtSYKkHOE-aGTDPcvs?usp=sharing">here</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">script</span> echo &quot;Skip cell segmentation&quot;

wsi.process(method=&#39;cellpose&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Skip cell segmentation
</pre></div></div>
</div>
<p>Extract handcrafted image features based on the image and nuclei segmentation result.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cell_mask_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s1">&#39;WSI_DCIS_10x&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_mask.npz&#39;</span><span class="p">)</span>
<span class="n">wsi</span> <span class="o">=</span> <span class="n">WholeSlideImage</span><span class="p">(</span><span class="n">image_path</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">sn</span><span class="p">,</span> <span class="n">nuclei_seg_path</span><span class="o">=</span><span class="n">cell_mask_path</span><span class="p">,</span> <span class="n">nuclei_seg_format</span><span class="o">=</span><span class="s1">&#39;mask_array_npz&#39;</span><span class="p">)</span>
<span class="n">wsi</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-cyan-fg">[12/13/23 16:53:37]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: <span class="ansi-cyan-intense-fg ansi-bold">76396</span> cells detected.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: <span class="ansi-cyan-intense-fg ansi-bold">76396</span> cells mapped to spots.
<span class="ansi-cyan-fg">[12/13/23 16:53:38]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: <span class="ansi-cyan-intense-fg ansi-bold">76359</span> cells kept.
&lt;function mean at 0x107c40900&gt;
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Extracting image features from a square image patch of size <span class="ansi-cyan-intense-fg ansi-bold">112</span> pixels centered at the cell.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Extracting gray image features: 100%|██████████| 76359/76359 [00:04&lt;00:00, 16574.04it/s]
Extracting RGB features: 100%|██████████| 76359/76359 [00:33&lt;00:00, 2297.83it/s]
</pre></div></div>
</div>
<p>Preprocessing spatial transcriptome. We used SCANPY to generate an adata for the spots including QC.</p>
<p>We follow the standarized steps used by SCANPY to create the spot adata from the space ranger output directory (<code class="docutils literal notranslate"><span class="pre">visium_dir</span></code>). The spot adata contains the expression matrix, the location of the spots as pixel positions on the WSI, and the hires, lowres images with scalefactors. - spot.h5ad</p>
<p>We skip this step here and the files can be downloaded from <a class="reference external" href="https://drive.google.com/drive/folders/1nX6c8FJqj-VCiCa9G4PM2vk-Ogp-9vEH?usp=sharing">here</a></p>
</section>
<section id="Predicting-cell-level-gene-expression-using-Thor-diffusion">
<h3>Predicting cell-level gene expression using Thor diffusion<a class="headerlink" href="#Predicting-cell-level-gene-expression-using-Thor-diffusion" title="Link to this heading"></a></h3>
<p>After finishing the preprocessing, you should have those files: - The original WSI (image_path) - The cell(nuclei) mask and features (in directory “./WSI_DCIS_10x”) - The spot-level gene expression (in directory “./Spatial_DCIS_10x”)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">outdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>

<span class="n">image_process_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="s2">&quot;WSI_DCIS_10x&quot;</span><span class="p">)</span>
<span class="n">cell_mask_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">image_process_dir</span><span class="p">,</span> <span class="s2">&quot;cell_mask.npz&quot;</span><span class="p">)</span>
<span class="n">cell_feature_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">image_process_dir</span><span class="p">,</span> <span class="s2">&quot;cell_features.csv&quot;</span><span class="p">)</span>

<span class="n">spatial_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="s2">&quot;Spatial_DCIS_10x&quot;</span><span class="p">)</span>
<span class="n">spot_adata_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">spatial_dir</span><span class="p">,</span> <span class="s2">&quot;spot.h5ad&quot;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<p>The first step is to map the spot gene expression to the segmented cells. We use the nearest neighbors approach. This cell-level gene expression is the starting point for the diffusion process.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">DCIS</span> <span class="o">=</span> <span class="n">fineST</span><span class="p">(</span>
    <span class="n">image_path</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s2">&quot;DCIS_10x&quot;</span><span class="p">,</span>
    <span class="n">spot_adata_path</span><span class="o">=</span><span class="n">spot_adata_path</span><span class="p">,</span>
    <span class="n">cell_features_csv_path</span><span class="o">=</span><span class="n">cell_feature_path</span>
<span class="p">)</span>

<span class="n">DCIS</span><span class="o">.</span><span class="n">prepare_input</span><span class="p">(</span><span class="n">mapping_margin</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-cyan-fg">[12/13/23 16:54:25]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Please check alignment of cells and spots
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: The first two columns in the node_feat DataFrame need to match the spatial coordinates from obsm<span class="ansi-bold">[</span><span class="ansi-green-fg">&#39;spatial&#39;</span><span class="ansi-bold">]</span>.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Mapping cells to the closest spots within <span class="ansi-cyan-intense-fg ansi-bold">10</span> x the spot radius
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_18_1.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_18_1.png" style="width: 947px; height: 854px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">DCIS</span><span class="o">.</span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 76344 × 16784
    obs: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;, &#39;n_genes&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;spot_barcodes&#39;, &#39;x&#39;, &#39;y&#39;, &#39;mean_gray&#39;, &#39;std_gray&#39;, &#39;entropy_img&#39;, &#39;mean_r&#39;, &#39;mean_g&#39;, &#39;mean_b&#39;, &#39;std_r&#39;, &#39;std_g&#39;, &#39;std_b&#39;, &#39;spot_heterogeneity&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;mt&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;
    uns: &#39;spatial&#39;, &#39;cell_image_props&#39;
    obsm: &#39;spatial&#39;
</pre></div></div>
</div>
<p>There are 75117 cells and for the sake of time, we’ll show prediction of a few genes. The same Markov transition matrix can be applied to all genes. The user-defined genes can be input by either in a 1-column text file or directly as an attribute of the DCIS object</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">DCIS</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;VEGFA&#39;</span><span class="p">,</span> <span class="s1">&#39;FGFR4&#39;</span><span class="p">,</span> <span class="s1">&#39;TPD52&#39;</span><span class="p">,</span> <span class="s1">&#39;GRB7&#39;</span><span class="p">,</span> <span class="s1">&#39;JUP&#39;</span><span class="p">,</span> <span class="s1">&#39;SCGB2A2&#39;</span><span class="p">,</span> <span class="s1">&#39;KANK1&#39;</span><span class="p">,</span> <span class="s1">&#39;ESR1&#39;</span><span class="p">,</span> <span class="s1">&#39;TFRC&#39;</span><span class="p">,</span> <span class="s1">&#39;ERBB2&#39;</span><span class="p">]</span>

<span class="n">DCIS</span><span class="o">.</span><span class="n">set_genes_for_prediction</span><span class="p">(</span><span class="n">genes_selection_key</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">DCIS</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">DCIS</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;used_for_prediction&#39;</span><span class="p">]]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_ids</th>
      <th>feature_types</th>
      <th>genome</th>
      <th>n_cells</th>
      <th>mt</th>
      <th>n_cells_by_counts</th>
      <th>mean_counts</th>
      <th>pct_dropout_by_counts</th>
      <th>total_counts</th>
      <th>used_for_prediction</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>TFRC</th>
      <td>ENSG00000072274</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>1847</td>
      <td>False</td>
      <td>1847</td>
      <td>2.779190</td>
      <td>26.648133</td>
      <td>6998.0</td>
      <td>True</td>
    </tr>
    <tr>
      <th>FGFR4</th>
      <td>ENSG00000160867</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>165</td>
      <td>False</td>
      <td>165</td>
      <td>0.094122</td>
      <td>93.447180</td>
      <td>237.0</td>
      <td>True</td>
    </tr>
    <tr>
      <th>VEGFA</th>
      <td>ENSG00000112715</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>1900</td>
      <td>False</td>
      <td>1900</td>
      <td>3.851866</td>
      <td>24.543288</td>
      <td>9699.0</td>
      <td>True</td>
    </tr>
    <tr>
      <th>ESR1</th>
      <td>ENSG00000091831</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>166</td>
      <td>False</td>
      <td>166</td>
      <td>0.069102</td>
      <td>93.407466</td>
      <td>174.0</td>
      <td>True</td>
    </tr>
    <tr>
      <th>TPD52</th>
      <td>ENSG00000076554</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>2016</td>
      <td>False</td>
      <td>2016</td>
      <td>5.621922</td>
      <td>19.936458</td>
      <td>14156.0</td>
      <td>True</td>
    </tr>
    <tr>
      <th>KANK1</th>
      <td>ENSG00000107104</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>818</td>
      <td>False</td>
      <td>818</td>
      <td>0.499603</td>
      <td>67.513900</td>
      <td>1258.0</td>
      <td>True</td>
    </tr>
    <tr>
      <th>SCGB2A2</th>
      <td>ENSG00000110484</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>2213</td>
      <td>False</td>
      <td>2213</td>
      <td>17.403097</td>
      <td>12.112788</td>
      <td>43821.0</td>
      <td>True</td>
    </tr>
    <tr>
      <th>ERBB2</th>
      <td>ENSG00000141736</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>2417</td>
      <td>False</td>
      <td>2417</td>
      <td>22.986498</td>
      <td>4.011120</td>
      <td>57880.0</td>
      <td>True</td>
    </tr>
    <tr>
      <th>GRB7</th>
      <td>ENSG00000141738</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>1411</td>
      <td>False</td>
      <td>1411</td>
      <td>2.008737</td>
      <td>43.963463</td>
      <td>5058.0</td>
      <td>True</td>
    </tr>
    <tr>
      <th>JUP</th>
      <td>ENSG00000173801</td>
      <td>Gene Expression</td>
      <td>GRCh38</td>
      <td>1758</td>
      <td>False</td>
      <td>1758</td>
      <td>2.829230</td>
      <td>30.182685</td>
      <td>7124.0</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">DCIS</span><span class="o">.</span><span class="n">recipe</span> <span class="o">=</span> <span class="s1">&#39;gene&#39;</span>
<span class="n">DCIS</span><span class="o">.</span><span class="n">set_params</span><span class="p">(</span>
    <span class="n">is_rawCount</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">out_prefix</span><span class="o">=</span><span class="s2">&quot;fineST&quot;</span><span class="p">,</span>
    <span class="n">write_freq</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">n_iter</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span>
    <span class="n">conn_csr_matrix</span><span class="o">=</span><span class="s2">&quot;force&quot;</span><span class="p">,</span>
    <span class="n">smoothing_scale</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
    <span class="n">node_features_obs_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;spot_heterogeneity&#39;</span><span class="p">],</span>
    <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">geom_morph_ratio</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">geom_constraint</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">inflation_percentage</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">regulate_expression_mean</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">stochastic_expression_neighbors_level</span><span class="o">=</span><span class="s1">&#39;spot&#39;</span><span class="p">,</span>
    <span class="n">smooth_predicted_expression_steps</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">reduced_dimension_transcriptome_obsm_key</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">adjust_cell_network_by_transcriptome_scale</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">n_jobs</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">DCIS</span><span class="o">.</span><span class="n">predict_gene_expression</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-cyan-fg">[12/13/23 16:54:27]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Using mode gene
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Forcing to recalculate the connectivities.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Construct SNN with morphological features: <span class="ansi-bold">[</span><span class="ansi-green-fg">&#39;mean_gray&#39;</span>, <span class="ansi-green-fg">&#39;std_gray&#39;</span>, <span class="ansi-green-fg">&#39;entropy_img&#39;</span>, <span class="ansi-green-fg">&#39;mean_r&#39;</span>, <span class="ansi-green-fg">&#39;mean_g&#39;</span>, <span class="ansi-green-fg">&#39;mean_b&#39;</span>,
<span class="ansi-cyan-fg">                    </span>         <span class="ansi-green-fg">&#39;std_r&#39;</span>, <span class="ansi-green-fg">&#39;std_g&#39;</span>, <span class="ansi-green-fg">&#39;std_b&#39;</span><span class="ansi-bold">]</span>.
<span class="ansi-cyan-fg">[12/13/23 16:54:28]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Finish constructing SNN
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Add adata.obsp<span class="ansi-bold">[</span><span class="ansi-green-fg">&#34;snn_connectivities&#34;</span><span class="ansi-bold">]</span>
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Add adata.obsp<span class="ansi-bold">[</span><span class="ansi-green-fg">&#34;snn_knn_connectivities&#34;</span><span class="ansi-bold">]</span>
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Add adata.uns<span class="ansi-bold">[</span><span class="ansi-green-fg">&#34;snn&#34;</span><span class="ansi-bold">]</span>
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Weigh cells according to the spot heterogeneity.
<span class="ansi-cyan-fg">[12/13/23 16:54:29]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Promote flow of information from more homogeneous cells to less.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Eliminate low quality edges <span class="ansi-bold">(</span>&lt;<span class="ansi-cyan-intense-fg ansi-bold">0.1</span><span class="ansi-bold">)</span> between cells.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Compute transition matrix
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: self weight scale is set to: <span class="ansi-cyan-intense-fg ansi-bold">0.200</span>
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Added transition matrix to adata.obsp<span class="ansi-bold">[</span><span class="ansi-green-fg">&#34;snn_transition_matrix&#34;</span><span class="ansi-bold">]</span>
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;spot_barcodes&#39; as categorical
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-cyan-fg">[12/13/23 16:54:30]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: fineST estimation starts.
<span class="ansi-cyan-fg">[12/13/23 16:54:48]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: fineST estimation finished.
<span class="ansi-cyan-fg">                    </span>
<span class="ansi-cyan-fg">                    </span>
</pre></div></div>
</div>
<p>Check the expression of a few genes.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad_thor</span> <span class="o">=</span> <span class="n">DCIS</span><span class="o">.</span><span class="n">load_result</span><span class="p">(</span><span class="s1">&#39;fineST_20_samp1.npz&#39;</span><span class="p">)</span>
<span class="n">ad_thor</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 76344 × 10
    obs: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;, &#39;n_genes&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;spot_barcodes&#39;, &#39;x&#39;, &#39;y&#39;, &#39;mean_gray&#39;, &#39;std_gray&#39;, &#39;entropy_img&#39;, &#39;mean_r&#39;, &#39;mean_g&#39;, &#39;mean_b&#39;, &#39;std_r&#39;, &#39;std_g&#39;, &#39;std_b&#39;, &#39;spot_heterogeneity&#39;, &#39;node_weights&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;mt&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;used_for_prediction&#39;, &#39;used_for_reduced&#39;, &#39;used_for_vae&#39;
    uns: &#39;spatial&#39;, &#39;cell_image_props&#39;, &#39;snn&#39;
    obsm: &#39;spatial&#39;
    obsp: &#39;snn_connectivities&#39;, &#39;snn_knn_connectivities&#39;, &#39;snn_transition_matrix&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">genes</span> <span class="o">=</span> <span class="n">DCIS</span><span class="o">.</span><span class="n">genes</span>

<span class="n">ad_spot</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">spot_adata_path</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">ad_spot</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="s1">&#39;p99&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_27_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_27_0.png" style="width: 1249px; height: 846px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">genes</span> <span class="o">=</span> <span class="n">DCIS</span><span class="o">.</span><span class="n">genes</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">ad_thor</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="s1">&#39;p99&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_28_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_28_0.png" style="width: 1250px; height: 846px;" />
</div>
</div>
</section>
</section>
<section id="Advance-analysis">
<h2>Advance analysis<a class="headerlink" href="#Advance-analysis" title="Link to this heading"></a></h2>
<p>In this notebook, we’ll perform analyses on the Ductal Carcinoma In Situ (DCIS) spatial data at cell-level inferred by Thor, as follows,</p>
<ul class="simple">
<li><p>Visualization of genes in the physical context at cell level</p></li>
<li><p>Tumor activity using oncogenes and tumor suppressor genes</p></li>
<li><p>Hallmark pathway enrichment</p></li>
<li><p>Copy number variation calculation</p></li>
<li><p>Tertiary Lymphoid Structures quantification</p></li>
</ul>
<p>For installation of Thor, please refer to Thor website.</p>
<p>The cell-level expression of the highly variable genes (2747 genes) inferred by Thor can be downloaded from the <a class="reference external" href="https://drive.google.com/drive/folders/18Wu1k09nQQ7gK40oRL5AWHDy-bDBc9en?usp=sharing">link</a>. You can also download the nuclei segmentation masks and full resolution image in the same link if you skip first part of the notebooks in this case study.</p>
<section id="Import-packages">
<h3>Import packages<a class="headerlink" href="#Import-packages" title="Link to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">scipy</span> <span class="kn">import</span> <span class="n">stats</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>
<span class="kn">from</span> <span class="nn">PIL</span> <span class="kn">import</span> <span class="n">Image</span>
<span class="kn">from</span> <span class="nn">shapely.geometry</span> <span class="kn">import</span> <span class="n">Polygon</span>


<span class="kn">import</span> <span class="nn">thor</span>
<span class="kn">from</span> <span class="nn">thor.utils</span> <span class="kn">import</span> <span class="n">convert_pixel_to_micron_visium</span><span class="p">,</span> <span class="n">get_ROI_tuple_from_polygon</span><span class="p">,</span> <span class="n">get_adata_layer_array</span>
<span class="kn">from</span> <span class="nn">thor.pl</span> <span class="kn">import</span> <span class="n">annotate_ROI</span><span class="p">,</span> <span class="n">get_nuclei_pixels</span><span class="p">,</span> <span class="n">single_molecule</span>
<span class="kn">from</span> <span class="nn">thor.analy</span> <span class="kn">import</span> <span class="n">get_pathway_score</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">config</span> InlineBackend.figure_format = &#39;retina&#39;
<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">scanpy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">dpi_save</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="s2">&quot;error&quot;</span>
</pre></div>
</div>
</div>
</section>
<section id="Load-the-image-(for-visualization)-and-Thor-predicted-cell-level-transcriptome-result">
<h3>Load the image (for visualization) and Thor-predicted cell-level transcriptome result<a class="headerlink" href="#Load-the-image-(for-visualization)-and-Thor-predicted-cell-level-transcriptome-result" title="Link to this heading"></a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">image_path</span> <span class="o">=</span> <span class="s2">&quot;Visium_FFPE_Human_Breast_Cancer_image.tif&quot;</span>
<span class="n">fullres_im</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">Image</span><span class="o">.</span><span class="n">open</span><span class="p">(</span><span class="n">image_path</span><span class="p">))</span>

<span class="n">adata_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;Thor_DCIS_10x&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_data_thor.h5ad&quot;</span><span class="p">)</span>
<span class="n">ad</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">adata_path</span><span class="p">)</span>
<span class="n">ad</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 76320 × 2748
    obs: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;, &#39;n_genes&#39;, &#39;n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;spot_barcodes&#39;, &#39;x&#39;, &#39;y&#39;, &#39;mean_gray&#39;, &#39;std_gray&#39;, &#39;entropy_img&#39;, &#39;mean_r&#39;, &#39;mean_g&#39;, &#39;mean_b&#39;, &#39;std_r&#39;, &#39;std_g&#39;, &#39;std_b&#39;, &#39;spot_heterogeneity&#39;, &#39;node_weights&#39;, &#39;tumor_region&#39;, &#39;tumor_region_ext&#39;, &#39;copy_variation&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;n_cells&#39;, &#39;mt&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;used_for_prediction&#39;, &#39;used_for_reduced&#39;, &#39;used_for_vae&#39;
    uns: &#39;cell_image_props&#39;, &#39;copy_variation_colors&#39;, &#39;hvg&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;recluster_normal_colors&#39;, &#39;snn&#39;, &#39;spatial&#39;, &#39;tumor_cluster_CNV_colors&#39;, &#39;tumor_normal_subtumor_colors&#39;, &#39;tumor_region_boundaries&#39;, &#39;umap&#39;
    obsm: &#39;spatial&#39;
    obsp: &#39;snn_connectivities&#39;, &#39;snn_knn_connectivities&#39;, &#39;snn_transition_matrix&#39;
</pre></div></div>
</div>
</section>
<section id="Visualize-genes-at-tissue-and-cell-level">
<h3>Visualize genes at tissue and cell level<a class="headerlink" href="#Visualize-genes-at-tissue-and-cell-level" title="Link to this heading"></a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Show all the tumor regions</span>
<span class="c1"># &#39;0&#39; means non-tumor</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">ad</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;tumor_region&#39;</span><span class="p">,</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;tab20b&#39;</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s1">&#39;all major tumor regions&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_38_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_38_0.png" style="width: 381px; height: 296px;" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">palette</span> <span class="o">=</span> <span class="n">thor</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">continuous_palettes</span><span class="p">[</span><span class="s1">&#39;blueYellow&#39;</span><span class="p">]</span>
<span class="n">cmap</span> <span class="o">=</span> <span class="n">mpl</span><span class="o">.</span><span class="n">colors</span><span class="o">.</span><span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span><span class="s1">&#39;my_cmap&#39;</span><span class="p">,</span> <span class="n">palette</span><span class="p">,</span> <span class="n">N</span><span class="o">=</span><span class="mi">256</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene</span> <span class="o">=</span> <span class="s1">&#39;VEGFA&#39;</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">ad</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;VEGFA&#39;</span><span class="p">,</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="s1">&#39;p99&#39;</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_40_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_40_0.png" style="width: 302px; height: 281px;" />
</div>
</div>
<p>Close-up gene expression in cells can be seamlessly explored using our dedicated webapp <a class="reference internal" href="../Mjolnir_installation.html"><span class="doc">Mjolnir</span></a> or as static figures plotted using Thor’s plotting functions. Here we annotate tumor regions on the H &amp; E image according to Agoko NV, Belgium on <a class="reference external" href="https://cf.10xgenomics.com/samples/spatial-exp/1.3.0/Visium_FFPE_Human_Breast_Cancer/Visium_FFPE_Human_Breast_Cancer_Pathologist_Annotations.png">10x Genomics website</a>.</p>
<p>We then use <em>Mjolnir</em> to extract the tumor regions and gene expression matrices of the cells in those regions.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tumor1</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;tumor_region_boundaries&#39;</span><span class="p">][</span><span class="s1">&#39;1&#39;</span><span class="p">]</span>
<span class="n">tumor1_polygon</span> <span class="o">=</span> <span class="n">Polygon</span><span class="p">(</span><span class="n">tumor1</span><span class="p">)</span>
<span class="n">ROI_tuple</span> <span class="o">=</span> <span class="n">get_ROI_tuple_from_polygon</span><span class="p">(</span><span class="n">tumor1</span><span class="p">,</span> <span class="n">extend_pixels</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>
<span class="n">annotate_ROI</span><span class="p">(</span><span class="n">fullres_im</span><span class="p">,</span> <span class="n">ROI_polygon</span><span class="o">=</span><span class="n">tumor1_polygon</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_42_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_42_0.png" style="width: 265px; height: 240px;" />
</div>
</div>
<p>load the cellmasks</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">cells_pixels</span> <span class="o">=</span> <span class="n">get_nuclei_pixels</span><span class="p">(</span><span class="n">ad</span><span class="p">,</span> <span class="s2">&quot;WSI_DCIS_10x/cell_mask.npz&quot;</span><span class="p">)</span>
<span class="n">microPerPixel</span> <span class="o">=</span> <span class="n">convert_pixel_to_micron_visium</span><span class="p">(</span><span class="n">ad</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gene</span> <span class="o">=</span> <span class="s1">&#39;VEGFA&#39;</span>
<span class="n">expression_vector</span> <span class="o">=</span> <span class="n">get_adata_layer_array</span><span class="p">(</span><span class="n">ad</span><span class="p">[:,</span> <span class="n">gene</span><span class="p">])</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">single_molecule</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">expression_vector</span><span class="p">,</span> <span class="n">cells_pixels</span><span class="p">,</span> <span class="n">full_res_im</span><span class="o">=</span><span class="n">fullres_im</span><span class="p">,</span> <span class="n">ROI_tuple</span><span class="o">=</span><span class="n">ROI_tuple</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">img_alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="s1">&#39;p99&#39;</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">500</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">2</span><span class="p">))</span>
<span class="n">thor</span><span class="o">.</span><span class="n">plotting</span><span class="o">.</span><span class="n">fill</span><span class="o">.</span><span class="n">add_scalebar</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">microPerPixel</span><span class="p">,</span> <span class="s1">&#39;um&#39;</span><span class="p">,</span> <span class="s1">&#39;lower left&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_45_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_45_0.png" style="width: 1376px; height: 1057px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ROI_squares</span> <span class="o">=</span> <span class="p">[[</span><span class="mi">9000</span><span class="p">,</span> <span class="mi">7900</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">500</span><span class="p">],</span> <span class="p">[</span><span class="mi">10400</span><span class="p">,</span> <span class="mi">8900</span><span class="p">,</span> <span class="mi">500</span><span class="p">,</span> <span class="mi">500</span><span class="p">]]</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>

<span class="n">i</span> <span class="o">=</span> <span class="mi">1</span>
<span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">ROI_squares</span><span class="p">:</span>
    <span class="n">single_molecule</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">expression_vector</span><span class="p">,</span> <span class="n">cells_pixels</span><span class="p">,</span> <span class="n">full_res_im</span><span class="o">=</span><span class="n">fullres_im</span><span class="p">,</span> <span class="n">ROI_tuple</span><span class="o">=</span><span class="n">r</span><span class="p">,</span> <span class="n">img_alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">show_cbar</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">return_fig</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">single_molecule</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">expression_vector</span><span class="p">,</span> <span class="n">cells_pixels</span><span class="p">,</span> <span class="n">full_res_im</span><span class="o">=</span><span class="n">fullres_im</span><span class="p">,</span> <span class="n">ROI_tuple</span><span class="o">=</span><span class="n">r</span><span class="p">,</span> <span class="n">img_alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">200</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">show_cbar</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">return_fig</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">cmap</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="s1">&#39;p99&#39;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
    <span class="n">thor</span><span class="o">.</span><span class="n">plotting</span><span class="o">.</span><span class="n">fill</span><span class="o">.</span><span class="n">add_scalebar</span><span class="p">(</span><span class="n">ax</span><span class="p">,</span> <span class="n">microPerPixel</span><span class="p">,</span> <span class="s1">&#39;um&#39;</span><span class="p">,</span> <span class="s1">&#39;lower left&#39;</span><span class="p">)</span>
    <span class="n">i</span> <span class="o">=</span> <span class="n">i</span> <span class="o">-</span> <span class="mi">1</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_46_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_46_0.png" style="width: 262px; height: 277px;" />
</div>
</div>
</section>
<section id="Expression-profile-of-oncogene-and-tumor-suppressor-gene">
<h3>Expression profile of oncogene and tumor suppressor gene<a class="headerlink" href="#Expression-profile-of-oncogene-and-tumor-suppressor-gene" title="Link to this heading"></a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">vector_friendly</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">))</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">ad</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;ERBB2&quot;</span><span class="p">,</span> <span class="s2">&quot;ATM&quot;</span><span class="p">],</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">img_key</span><span class="o">=</span><span class="s2">&quot;hires&quot;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&lt;Axes: title={&#39;center&#39;: &#39;ERBB2&#39;}, xlabel=&#39;spatial1&#39;, ylabel=&#39;spatial2&#39;&gt;,
 &lt;Axes: title={&#39;center&#39;: &#39;ATM&#39;}, xlabel=&#39;spatial1&#39;, ylabel=&#39;spatial2&#39;&gt;]
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_48_1.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_48_1.png" style="width: 889px; height: 406px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">g1</span> <span class="o">=</span><span class="s2">&quot;ERBB2&quot;</span>
<span class="n">g2</span> <span class="o">=</span> <span class="s2">&quot;ATM&quot;</span>
<span class="n">region_col</span> <span class="o">=</span> <span class="s2">&quot;tumor_region&quot;</span>
<span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;7&quot;</span><span class="p">,</span> <span class="s2">&quot;1&quot;</span><span class="p">,</span> <span class="s2">&quot;17&quot;</span><span class="p">,</span> <span class="s2">&quot;11&quot;</span><span class="p">,</span> <span class="s2">&quot;6&quot;</span><span class="p">,</span> <span class="s2">&quot;15&quot;</span><span class="p">]:</span>

    <span class="n">x</span> <span class="o">=</span> <span class="n">ad</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">region_col</span><span class="p">]</span><span class="o">==</span><span class="n">region</span><span class="p">,</span> <span class="n">g1</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>
    <span class="n">y</span> <span class="o">=</span> <span class="n">ad</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">region_col</span><span class="p">]</span><span class="o">==</span><span class="n">region</span><span class="p">,</span> <span class="n">g2</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>

    <span class="n">values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">])</span>
    <span class="n">kernel</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">gaussian_kde</span><span class="p">(</span><span class="n">values</span><span class="p">)(</span><span class="n">values</span><span class="p">)</span>
    <span class="n">ax1</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">jointplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s2">&quot;kde&quot;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">levels</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

    <span class="n">ax1</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">cla</span><span class="p">()</span>
    <span class="n">plt</span><span class="o">.</span><span class="n">sca</span><span class="p">(</span><span class="n">ax1</span><span class="o">.</span><span class="n">ax_joint</span><span class="p">)</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">kernel</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;jet&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="o">.</span><span class="n">ax_joint</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">g1</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">g2</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_49_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_49_0.png" style="width: 492px; height: 486px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_49_1.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_49_1.png" style="width: 492px; height: 486px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_49_2.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_49_2.png" style="width: 492px; height: 486px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_49_3.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_49_3.png" style="width: 492px; height: 486px;" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_49_4.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_49_4.png" style="width: 492px; height: 486px;" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_49_5.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_49_5.png" style="width: 492px; height: 486px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">g1</span> <span class="o">=</span><span class="s2">&quot;ERBB2&quot;</span>
<span class="n">g2</span> <span class="o">=</span> <span class="s2">&quot;ATM&quot;</span>

<span class="n">x</span> <span class="o">=</span> <span class="n">ad</span><span class="p">[:,</span> <span class="n">g1</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">ad</span><span class="p">[:,</span> <span class="n">g2</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>

<span class="n">values</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">vstack</span><span class="p">([</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">])</span>
<span class="n">kernel</span> <span class="o">=</span> <span class="n">stats</span><span class="o">.</span><span class="n">gaussian_kde</span><span class="p">(</span><span class="n">values</span><span class="p">)(</span><span class="n">values</span><span class="p">)</span>
<span class="n">ax1</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">jointplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span> <span class="n">kind</span><span class="o">=</span><span class="s2">&quot;kde&quot;</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">levels</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span>

<span class="n">ax1</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">cla</span><span class="p">()</span>
<span class="n">plt</span><span class="o">.</span><span class="n">sca</span><span class="p">(</span><span class="n">ax1</span><span class="o">.</span><span class="n">ax_joint</span><span class="p">)</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">scatterplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">y</span><span class="p">,</span> <span class="n">s</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">kernel</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s2">&quot;jet&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="o">.</span><span class="n">ax_joint</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="n">g1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="n">g2</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mf">2.5</span><span class="p">)</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_50_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_50_0.png" style="width: 486px; height: 486px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">g1</span> <span class="o">=</span><span class="s2">&quot;ERBB2&quot;</span>
<span class="n">g2</span> <span class="o">=</span> <span class="s2">&quot;ATM&quot;</span>
<span class="n">region_col</span> <span class="o">=</span> <span class="s2">&quot;tumor_region&quot;</span>

<span class="n">x</span> <span class="o">=</span> <span class="n">ad</span><span class="p">[:,</span> <span class="n">g1</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>
<span class="n">y</span> <span class="o">=</span> <span class="n">ad</span><span class="p">[:,</span> <span class="n">g2</span><span class="p">]</span><span class="o">.</span><span class="n">X</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">]</span>

<span class="n">ad</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;severe&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">log2</span><span class="p">(</span><span class="n">x</span><span class="o">/</span><span class="p">(</span><span class="n">y</span><span class="o">+</span><span class="mf">0.01</span><span class="p">))</span>

<span class="c1"># tumor_region &#39;0&#39; is the background (non-tumor) region</span>
<span class="n">ad1</span> <span class="o">=</span> <span class="n">ad</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">region_col</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;0&#39;</span><span class="p">]</span>

<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">color_map</span><span class="o">=</span><span class="s2">&quot;RdYlGn_r&quot;</span><span class="p">,</span> <span class="n">vector_friendly</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="mi">18</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span><span class="mi">8</span><span class="p">))</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">ad1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;severe&quot;</span><span class="p">],</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=-</span><span class="mi">3</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">alpha_img</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_51_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_51_0.png" style="width: 522px; height: 520px;" />
</div>
</div>
</section>
<section id="Hallmark-pathway-enrichment">
<h3>Hallmark pathway enrichment<a class="headerlink" href="#Hallmark-pathway-enrichment" title="Link to this heading"></a></h3>
<p>Pathway enrichment analyses are done using decoupler-py.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">decoupler</span> <span class="k">as</span> <span class="nn">dc</span>

<span class="c1"># Load MSigDB. This can take a while depending on the internet connection (normally 1-2 minutes)</span>
<span class="n">msigdb</span> <span class="o">=</span> <span class="n">dc</span><span class="o">.</span><span class="n">get_resource</span><span class="p">(</span><span class="s1">&#39;MSigDB&#39;</span><span class="p">)</span>

<span class="c1"># Filter by hallmark</span>
<span class="n">msigdb</span> <span class="o">=</span> <span class="n">msigdb</span><span class="p">[</span><span class="n">msigdb</span><span class="p">[</span><span class="s1">&#39;collection&#39;</span><span class="p">]</span><span class="o">==</span><span class="s1">&#39;hallmark&#39;</span><span class="p">]</span>
<span class="n">msigdb</span> <span class="o">=</span> <span class="n">msigdb</span><span class="p">[</span><span class="o">~</span><span class="n">msigdb</span><span class="o">.</span><span class="n">duplicated</span><span class="p">([</span><span class="s1">&#39;geneset&#39;</span><span class="p">,</span> <span class="s1">&#39;genesymbol&#39;</span><span class="p">])]</span>

<span class="n">acts</span> <span class="o">=</span> <span class="n">get_pathway_score</span><span class="p">(</span><span class="n">ad</span><span class="p">,</span> <span class="n">layer</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">net_df</span><span class="o">=</span><span class="n">msigdb</span><span class="p">)</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">scale</span><span class="p">(</span><span class="n">acts</span><span class="p">,</span> <span class="n">max_value</span><span class="o">=</span><span class="mi">10</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
root - INFO - Downloading data from `https://omnipathdb.org/queries/enzsub?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/queries/interactions?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/queries/complexes?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/queries/annotations?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/queries/intercell?format=json`
root - INFO - Downloading data from `https://omnipathdb.org/about?format=text`
root - INFO - Downloading annotations for all proteins from the following resources: `[&#39;MSigDB&#39;]`
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Running ora on mat with 76320 samples and 2748 targets for 49 sources.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 76320/76320 [00:40&lt;00:00, 1864.33it/s]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[31]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">example_pathways</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;HALLMARK_DNA_REPAIR&#39;</span><span class="p">,</span> <span class="s1">&#39;HALLMARK_E2F_TARGETS&#39;</span><span class="p">,</span> <span class="s1">&#39;HALLMARK_ESTROGEN_RESPONSE_EARLY&#39;</span><span class="p">,</span> <span class="s1">&#39;HALLMARK_IL6_JAK_STAT3_SIGNALING&#39;</span><span class="p">]</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">acts</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">example_pathways</span><span class="p">,</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">alpha_img</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="s1">&#39;p99&#39;</span><span class="p">,</span> <span class="n">colorbar_loc</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_54_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_54_0.png" style="width: 1077px; height: 1086px;" />
</div>
</div>
</section>
<section id="Copy-number-variation">
<h3>Copy number variation<a class="headerlink" href="#Copy-number-variation" title="Link to this heading"></a></h3>
<p>This requires installation of CopyKAT and R environment in order to run the script. Here we simply show our result.</p>
<p><code class="docutils literal notranslate"><span class="pre">from</span> <span class="pre">thor.analy</span> <span class="pre">import</span> <span class="pre">prepare_and_run_copykat</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">prepare_and_run_copykat(adata,datadir,</span> <span class="pre">sam_name=&quot;BC&quot;)</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[32]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#sc.pl.spatial(ad, color=&#39;copy_variation&#39;, spot_size=50, ncols=1)</span>

<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">ad</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;copy_variation&#39;</span><span class="p">,</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="mi">4</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">img_key</span><span class="o">=</span><span class="s2">&quot;hires&quot;</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>

<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([])</span>

<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_56_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_56_0.png" style="width: 431px; height: 357px;" />
</div>
</div>
</section>
<section id="Tertiary-Lymphoid-Structures(TLS)-quantification">
<h3>Tertiary Lymphoid Structures(TLS) quantification<a class="headerlink" href="#Tertiary-Lymphoid-Structures(TLS)-quantification" title="Link to this heading"></a></h3>
<p>We use the list of 29 signature genes used by <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/35231421">this study</a>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[33]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># This is the gene list for the TLS score that we finally used, since it includes more related cell-type markers.</span>
<span class="n">TLS_list_immunity</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;IGHA1&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;IGHG1&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;IGHG2&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;IGHG3&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;IGHG4&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;IGHGP&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;IGHM&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;IGKC&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;IGLC1&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;IGLC2&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;IGLC3&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;JCHAIN&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;CD79A&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;FCRL5&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;MZB1&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;SSR4&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;XBP1&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;TRBC2&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;IL7R&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;CXCL12&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;LUM&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;C1QA&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;C7&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;CD52&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;APOE&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;PLTP&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;PTGDS&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;PIM2&#39;</span><span class="p">,</span>
                    <span class="s1">&#39;DERL3&#39;</span><span class="p">]</span>

<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">score_genes</span><span class="p">(</span><span class="n">ad</span><span class="p">,</span> <span class="n">TLS_list_immunity</span><span class="p">,</span> <span class="n">ctrl_size</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">TLS_list_immunity</span><span class="p">),</span> <span class="n">gene_pool</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">n_bins</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span> <span class="n">score_name</span><span class="o">=</span><span class="s1">&#39;TLS_score_immunity&#39;</span><span class="p">,</span> <span class="n">random_state</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">copy</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[34]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>


<span class="n">score</span> <span class="o">=</span> <span class="s1">&#39;TLS_score_immunity&#39;</span>
<span class="n">ad1</span> <span class="o">=</span> <span class="n">ad</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">score</span><span class="p">]</span><span class="o">&gt;</span><span class="mf">0.3</span><span class="p">]</span>

<span class="n">scalefactor</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">][</span><span class="s1">&#39;Visium_FFPE_Human_Breast_Cancer&#39;</span><span class="p">][</span><span class="s1">&#39;scalefactors&#39;</span><span class="p">][</span><span class="s1">&#39;tissue_hires_scalef&#39;</span><span class="p">]</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span><span class="mi">6</span><span class="p">))</span>

<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">ad1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">score</span><span class="p">,</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;viridis&#39;</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">colorbar_loc</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>

<span class="n">tumor_borders</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;tumor_region_boundaries&#39;</span><span class="p">]</span>
<span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">border</span> <span class="ow">in</span> <span class="n">tumor_borders</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">border</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="n">scalefactor</span><span class="p">,</span> <span class="n">border</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="n">scalefactor</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;r&#39;</span><span class="p">,</span> <span class="n">lw</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">zorder</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>


<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_59_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_59_0.png" style="width: 374px; height: 380px;" />
</div>
</div>
<p>Rank the tumor regions according to the TLS score. We extended the tumor regions by 1 spot (55 um) outwards.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[35]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">region_col</span> <span class="o">=</span> <span class="s1">&#39;tumor_region_ext&#39;</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="n">scores</span> <span class="o">=</span>  <span class="p">[</span><span class="s1">&#39;TLS_score_immunity&#39;</span><span class="p">]</span>

<span class="k">for</span> <span class="n">score</span> <span class="ow">in</span> <span class="n">scores</span><span class="p">:</span>
    <span class="n">ad1</span> <span class="o">=</span> <span class="n">ad</span><span class="p">[</span><span class="n">ad</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">score</span><span class="p">]</span><span class="o">&gt;</span><span class="mf">0.</span><span class="p">]</span>
    <span class="n">TLS_clone_stats1</span> <span class="o">=</span> <span class="n">ad1</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">region_col</span><span class="p">)[[</span><span class="n">score</span><span class="p">]]</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span><span class="s1">&#39;median&#39;</span><span class="p">)</span>
    <span class="n">rank</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">TLS_clone_stats1</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="n">by</span><span class="o">=</span><span class="n">score</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
    <span class="n">rank</span><span class="o">.</span><span class="n">remove</span><span class="p">(</span><span class="s1">&#39;0&#39;</span><span class="p">)</span>

    <span class="nb">print</span><span class="p">(</span><span class="n">rank</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;7&#39;, &#39;1&#39;, &#39;17&#39;, &#39;16&#39;, &#39;4&#39;, &#39;8&#39;, &#39;5&#39;, &#39;12&#39;, &#39;9&#39;, &#39;18&#39;, &#39;2&#39;, &#39;14&#39;, &#39;10&#39;, &#39;3&#39;, &#39;13&#39;, &#39;11&#39;, &#39;6&#39;, &#39;15&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">from</span> <span class="nn">matplotlib</span> <span class="kn">import</span> <span class="n">pyplot</span> <span class="k">as</span> <span class="n">plt</span>

<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;white&#39;</span><span class="p">)</span>

<span class="n">tumor_col</span> <span class="o">=</span> <span class="s1">&#39;tumor_region_ext&#39;</span>
<span class="n">score</span> <span class="o">=</span> <span class="s1">&#39;TLS_score_immunity&#39;</span>

<span class="n">df</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">obs</span><span class="p">[[</span><span class="n">region_col</span><span class="p">,</span> <span class="n">score</span><span class="p">]]</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">score</span><span class="p">]</span><span class="o">&gt;</span><span class="mi">0</span><span class="p">,</span> <span class="p">:]</span>
<span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="n">region_col</span><span class="p">]</span><span class="o">!=</span><span class="s1">&#39;0&#39;</span><span class="p">,</span> <span class="p">:]</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">6</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>

<span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="o">.</span><span class="n">right</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;none&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">spines</span><span class="o">.</span><span class="n">top</span><span class="o">.</span><span class="n">set_color</span><span class="p">(</span><span class="s1">&#39;none&#39;</span><span class="p">)</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">boxplot</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">score</span><span class="p">,</span> <span class="n">x</span><span class="o">=</span><span class="n">region_col</span><span class="p">,</span> <span class="n">whis</span><span class="o">=</span><span class="p">[</span><span class="mi">5</span><span class="p">,</span><span class="mi">95</span><span class="p">],</span> <span class="n">dodge</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">width</span><span class="o">=</span><span class="mf">0.3</span><span class="p">,</span> <span class="n">fliersize</span><span class="o">=</span><span class="mf">0.</span><span class="p">,</span> <span class="n">order</span><span class="o">=</span><span class="n">rank</span> <span class="p">)</span>
<span class="k">for</span> <span class="n">patch</span> <span class="ow">in</span> <span class="n">ax</span><span class="o">.</span><span class="n">patches</span><span class="p">:</span>
    <span class="n">r</span><span class="p">,</span> <span class="n">g</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="n">a</span> <span class="o">=</span> <span class="n">patch</span><span class="o">.</span><span class="n">get_facecolor</span><span class="p">()</span>
    <span class="n">patch</span><span class="o">.</span><span class="n">set_facecolor</span><span class="p">((</span><span class="n">r</span><span class="p">,</span> <span class="n">g</span><span class="p">,</span> <span class="n">b</span><span class="p">,</span> <span class="mf">.5</span><span class="p">))</span>
<span class="c1">#sns.stripplot(df, y=score, x=region_col, ax=ax, size=0.8, jitter=0.05, dodge=True, legend=&quot;&quot;, alpha=1)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">_remove_legend</span><span class="p">(</span><span class="n">ax</span><span class="o">.</span><span class="n">get_legend</span><span class="p">())</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mf">0.6</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;Tumor region&#39;</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;TLS score&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
matplotlib.category - INFO - Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
matplotlib.category - INFO - Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[36]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Text(0, 0.5, &#39;TLS score&#39;)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_DCIS_62_2.png" class="no-scaled-link" src="../_images/notebooks_runThor_DCIS_62_2.png" style="width: 438px; height: 366px;" />
</div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="runThor_MOB.html" class="btn btn-neutral float-left" title="Mouse olfactory bulb" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="run_thor_analyses_HF.html" class="btn btn-neutral float-right" title="Human heart failure" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Wang Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>