<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Mouse olfactory bulb &mdash; Thor  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="../_static/style.css?v=793f93ba" />
      <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="../_static/documentation_options.js?v=7f41d439"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
        <script src="../_static/copybutton.js?v=f281be69"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Ductal carcinoma in situ" href="runThor_DCIS.html" />
    <link rel="prev" title="thor.pl.fringe" href="../_autosummary/thor.pl.fringe.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            Thor
              <img src="../_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Thor</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Thor Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../release_notes.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../FAQ.html">Frequently asked questions</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../API.html">Classes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../API.html#api">API</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Mouse olfactory bulb</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Import-the-packages">Import the packages</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Preprocessing">Preprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Preprocessing-the-whole-slide-image,-including-cell-segmentation-and-morphological-feature-extraction.-If-not-specified,-the-segmentation-data-will-be-saved-in-a-directory-(created-by-Thor)-WSI_{sn}.">Preprocessing the whole slide image, including cell segmentation and morphological feature extraction. If not specified, the segmentation data will be saved in a directory (created by Thor) WSI_{sn}.</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Preprocessing-spatial-transcriptome.-We-used-SCANPY-to-generate-an-adata-for-the-spots-including-QC.">Preprocessing spatial transcriptome. We used SCANPY to generate an adata for the spots including QC.</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Predicting-cell-level-gene-expression-using-Markov-graph-diffusion">Predicting cell-level gene expression using Markov graph diffusion</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Check-the-expression-of-a-few-genes.">Check the expression of a few genes.</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="runThor_DCIS.html">Ductal carcinoma in situ</a></li>
<li class="toctree-l1"><a class="reference internal" href="run_thor_analyses_HF.html">Human heart failure</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_runThor_VisiumHD.html">Run Thor on Visium HD data</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial_run_cell_communication_commot.html">Cell cell communication</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Mjolnir visualization</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../Mjolnir.html">About Mjolnir</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Mjolnir_installation.html">Mjolnir Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Mjolnir_video_tutorials.html">Mjolnir video tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Mjolnir_test_data.html">Mjolnir test Data</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">Thor</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Mouse olfactory bulb</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <div class="admonition note">
This page was generated from a Jupyter notebook. You can download the notebook
<a href="https://github.com/GuangyuWangLab2021/Thor/blob/main/notebooks/runThor_MOB.ipynb" download>here</a> to run it locally in Jupyter.
</div><section id="Mouse-olfactory-bulb">
<h1>Mouse olfactory bulb<a class="headerlink" href="#Mouse-olfactory-bulb" title="Link to this heading"></a></h1>
<p>Thor reveals cell-resolution mouse olfactory bulb structure.</p>
<p>In this tutorial, we show inference of cell-level spatial gene expression from the Visium spot-level spatial transcriptome and a whole slide image (H&amp;E staining) using Thor.</p>
<p>The spatial transcriptome dataset is the Adult Mouse Olfactory Bulb. The input data can be downloaded from our <a class="reference external" href="https://drive.google.com/drive/folders/1CaUHMRUcKT-qGY2_lMkOcVVu_7OW50S-?usp=sharing">google drive</a> or 10x Genomics <a class="reference external" href="https://www.10xgenomics.com/resources/datasets/adult-mouse-olfactory-bulb-1-standard-1">website</a>.</p>
<p>For installation of Thor, please refer to <a class="reference internal" href="../installation.html"><span class="doc">this installation guide</span></a>.</p>
<section id="Import-the-packages">
<h2>Import the packages<a class="headerlink" href="#Import-the-packages" title="Link to this heading"></a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">datetime</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">()</span>
<span class="n">logger</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">)</span>
<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%(name)s</span><span class="s1"> - </span><span class="si">%(levelname)s</span><span class="s1"> - </span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="n">now</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span>
<span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Current Time: </span><span class="si">{</span><span class="n">now</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
root - INFO - Current Time: 2023-12-13 12:49:20.893931
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">config</span> InlineBackend.figure_format = &#39;retina&#39;

<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">matplotlib.colors</span> <span class="kn">import</span> <span class="n">LinearSegmentedColormap</span>

<span class="kn">import</span> <span class="nn">warnings</span>

<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">scanpy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">dpi_save</span><span class="o">=</span><span class="mi">300</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="s1">&#39;error&#39;</span>


<span class="kn">from</span> <span class="nn">thor.pp</span> <span class="kn">import</span> <span class="n">WholeSlideImage</span><span class="p">,</span> <span class="n">Spatial</span>
<span class="kn">from</span> <span class="nn">thor.finest</span> <span class="kn">import</span> <span class="n">fineST</span>
<span class="kn">from</span> <span class="nn">thor.pl</span> <span class="kn">import</span> <span class="n">colors</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
numexpr.utils - INFO - Note: NumExpr detected 12 cores but &#34;NUMEXPR_MAX_THREADS&#34; not set, so enforcing safe limit of 8.
numexpr.utils - INFO - NumExpr defaulting to 8 threads.
</pre></div></div>
</div>
</section>
<section id="Preprocessing">
<h2>Preprocessing<a class="headerlink" href="#Preprocessing" title="Link to this heading"></a></h2>
<section id="Preprocessing-the-whole-slide-image,-including-cell-segmentation-and-morphological-feature-extraction.-If-not-specified,-the-segmentation-data-will-be-saved-in-a-directory-(created-by-Thor)-WSI_{sn}.">
<h3>Preprocessing the whole slide image, including cell segmentation and morphological feature extraction. If not specified, the segmentation data will be saved in a directory (created by Thor) WSI_{sn}.<a class="headerlink" href="#Preprocessing-the-whole-slide-image,-including-cell-segmentation-and-morphological-feature-extraction.-If-not-specified,-the-segmentation-data-will-be-saved-in-a-directory-(created-by-Thor)-WSI_{sn}." title="Link to this heading"></a></h3>
<p>The WSI can be downloaded directly from <a class="reference external" href="https://drive.google.com/drive/folders/1CaUHMRUcKT-qGY2_lMkOcVVu_7OW50S-?usp=sharing">google drive</a>.</p>
<p>The outcomes of image preprocessing are two files: - cell mask - cell features</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sn</span> <span class="o">=</span> <span class="s1">&#39;MOB_10x&#39;</span>
<span class="n">image_path</span> <span class="o">=</span> <span class="s2">&quot;Visium_Mouse_Olfactory_Bulb_image.tif&quot;</span>

<span class="n">wsi</span> <span class="o">=</span> <span class="n">WholeSlideImage</span><span class="p">(</span><span class="n">image_path</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">sn</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Here we use Cellpose to segment the nuclei. Notice the segmentation can take a long time without GPU. The output file is a <code class="docutils literal notranslate"><span class="pre">n_pixel_row</span></code> x <code class="docutils literal notranslate"><span class="pre">n_pixel_col</span></code> matrix.</p>
<p>For the sake of reproducing the results in the paper, we’ll skip this step and load the pre-computed segmentation results from <a class="reference external" href="https://drive.google.com/drive/folders/1CaUHMRUcKT-qGY2_lMkOcVVu_7OW50S-?usp=sharing">here</a>.</p>
<p>In the downloaded sub-folder (“WSI_MOB_10x”), both the cell segmentation and cell features files are included.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%</span><span class="k">script</span> echo &quot;Skip cell segmentation&quot;

wsi.process(method=&#39;cellpose&#39;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Skip cell segmentation
</pre></div></div>
</div>
</section>
<section id="Preprocessing-spatial-transcriptome.-We-used-SCANPY-to-generate-an-adata-for-the-spots-including-QC.">
<h3>Preprocessing spatial transcriptome. We used SCANPY to generate an adata for the spots including QC.<a class="headerlink" href="#Preprocessing-spatial-transcriptome.-We-used-SCANPY-to-generate-an-adata-for-the-spots-including-QC." title="Link to this heading"></a></h3>
<p>We follow the standarized steps used by SCANPY to create the spot adata from the space ranger output directory (<code class="docutils literal notranslate"><span class="pre">st_dir</span></code>).</p>
<p>The spot adata contains the expression matrix, the location of the spots as pixel positions on the WSI, and the hires, lowres images with scalefactors. - spot.h5ad</p>
<p>You can skip this step and download the files <a class="reference external" href="https://drive.google.com/drive/folders/1CaUHMRUcKT-qGY2_lMkOcVVu_7OW50S-?usp=sharing">here</a>.The output is in the folder “Spatial_MOB_10x”.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">st_dir</span> <span class="o">=</span> <span class="s2">&quot;10xOlfactory&quot;</span>

<span class="n">trans</span> <span class="o">=</span> <span class="n">Spatial</span><span class="p">(</span><span class="n">sn</span><span class="p">,</span> <span class="n">st_dir</span><span class="p">)</span>
<span class="n">trans</span><span class="o">.</span><span class="n">process_transcriptome</span><span class="p">(</span><span class="n">perform_QC</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;feature_types&#39; as categorical
... storing &#39;genome&#39; as categorical
</pre></div></div>
</div>
<p>We can customize the transcriptome preprocessing if needed. Here we do log-normalization.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">spot_adata_path</span> <span class="o">=</span> <span class="n">trans</span><span class="o">.</span><span class="n">spot_adata_path</span>
<span class="n">adata_spot</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">spot_adata_path</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata_spot</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata_spot</span><span class="p">)</span>

<span class="n">adata_spot</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="n">spot_adata_path</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="Predicting-cell-level-gene-expression-using-Markov-graph-diffusion">
<h2>Predicting cell-level gene expression using Markov graph diffusion<a class="headerlink" href="#Predicting-cell-level-gene-expression-using-Markov-graph-diffusion" title="Link to this heading"></a></h2>
<p>After finishing the preprocessing, you should have those files: - The original WSI (<code class="docutils literal notranslate"><span class="pre">image_path</span></code>) - The cell(nuclei) mask and features (in directory “./WSI_MOB_10x”) - The spot-level gene expression (in directory “./Spatial_MOB_10x”)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">outdir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">getcwd</span><span class="p">()</span>

<span class="n">image_process_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">outdir</span><span class="p">,</span> <span class="s2">&quot;WSI_MOB_10x&quot;</span><span class="p">)</span>
<span class="n">cell_mask_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">image_process_dir</span><span class="p">,</span> <span class="s2">&quot;cell_mask.npz&quot;</span><span class="p">)</span>
<span class="n">cell_feature_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">image_process_dir</span><span class="p">,</span> <span class="s2">&quot;cell_features.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The first step is to map the spot gene expression to the segmented cells. We use the nearest neighbors approach. This cell-level gene expression is the starting point for the diffusion process.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">MOB</span> <span class="o">=</span> <span class="n">fineST</span><span class="p">(</span>
    <span class="n">image_path</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="n">sn</span><span class="p">,</span>
    <span class="n">spot_adata_path</span><span class="o">=</span><span class="n">spot_adata_path</span><span class="p">,</span>
    <span class="n">cell_features_csv_path</span><span class="o">=</span><span class="n">cell_feature_path</span><span class="p">,</span>
    <span class="n">cell_features_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;x&#39;</span><span class="p">,</span> <span class="s1">&#39;y&#39;</span><span class="p">,</span> <span class="s1">&#39;mean_gray&#39;</span><span class="p">,</span> <span class="s1">&#39;std_gray&#39;</span><span class="p">,</span> <span class="s1">&#39;entropy_img&#39;</span><span class="p">,</span> <span class="s1">&#39;mean_r&#39;</span><span class="p">,</span> <span class="s1">&#39;mean_g&#39;</span><span class="p">,</span> <span class="s1">&#39;mean_b&#39;</span><span class="p">,</span> <span class="s1">&#39;std_r&#39;</span><span class="p">,</span> <span class="s1">&#39;std_g&#39;</span><span class="p">,</span> <span class="s1">&#39;std_b&#39;</span><span class="p">],</span>
<span class="p">)</span>

<span class="n">MOB</span><span class="o">.</span><span class="n">prepare_input</span><span class="p">(</span><span class="n">mapping_margin</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-cyan-fg">[12/13/23 12:49:29]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Please check alignment of cells and spots
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: The first two columns in the node_feat DataFrame need to match the spatial coordinates from obsm<span class="ansi-bold">[</span><span class="ansi-green-fg">&#39;spatial&#39;</span><span class="ansi-bold">]</span>.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Mapping cells to the closest spots within <span class="ansi-cyan-intense-fg ansi-bold">2</span> x the spot radius
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_MOB_18_1.png" class="no-scaled-link" src="../_images/notebooks_runThor_MOB_18_1.png" style="width: 925px; height: 854px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">MOB</span><span class="o">.</span><span class="n">adata</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 40543 × 32285
    obs: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;, &#39;spot_barcodes&#39;, &#39;x&#39;, &#39;y&#39;, &#39;mean_gray&#39;, &#39;std_gray&#39;, &#39;entropy_img&#39;, &#39;mean_r&#39;, &#39;mean_g&#39;, &#39;mean_b&#39;, &#39;std_r&#39;, &#39;std_g&#39;, &#39;std_b&#39;, &#39;spot_heterogeneity&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;
    uns: &#39;log1p&#39;, &#39;spatial&#39;, &#39;cell_image_props&#39;
    obsm: &#39;spatial&#39;
</pre></div></div>
</div>
<p>Second step is set up genes for prediction. For the sake of time, we’ll show prediction of a few genes. The same Markov transition matrix can be applied to all genes. The user-defined genes can be input by either in a 1-column text file or directly as the attribute <code class="docutils literal notranslate"><span class="pre">genes</span></code> of the <code class="docutils literal notranslate"><span class="pre">MOB</span></code> object</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">MOB</span><span class="o">.</span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Eomes&#39;</span><span class="p">,</span> <span class="s1">&#39;Uchl1&#39;</span><span class="p">,</span> <span class="s1">&#39;Pcp4l1&#39;</span><span class="p">,</span> <span class="s1">&#39;Col6a1&#39;</span><span class="p">,</span> <span class="s1">&#39;Tbx21&#39;</span><span class="p">,</span> <span class="s1">&#39;Slc17a6&#39;</span><span class="p">,</span> <span class="s1">&#39;Bmp7&#39;</span><span class="p">,</span> <span class="s1">&#39;Camk2a&#39;</span><span class="p">,</span> <span class="s1">&#39;Cystm1&#39;</span><span class="p">,</span> <span class="s1">&#39;Penk&#39;</span><span class="p">,</span> <span class="s1">&#39;Gria1&#39;</span><span class="p">,</span> <span class="s1">&#39;Nap1l5&#39;</span><span class="p">,</span> <span class="s1">&#39;Serpine2&#39;</span><span class="p">,</span> <span class="s1">&#39;Kctd12&#39;</span><span class="p">,</span> <span class="s1">&#39;Pde5a&#39;</span><span class="p">,</span> <span class="s1">&#39;Syt7&#39;</span><span class="p">,</span> <span class="s1">&#39;Vtn&#39;</span><span class="p">]</span>
<span class="n">MOB</span><span class="o">.</span><span class="n">set_genes_for_prediction</span><span class="p">(</span><span class="n">genes_selection_key</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">MOB</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">MOB</span><span class="o">.</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s1">&#39;used_for_prediction&#39;</span><span class="p">]]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_ids</th>
      <th>feature_types</th>
      <th>genome</th>
      <th>used_for_prediction</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Serpine2</th>
      <td>ENSMUSG00000026249</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Pcp4l1</th>
      <td>ENSMUSG00000038370</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Bmp7</th>
      <td>ENSMUSG00000008999</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Pde5a</th>
      <td>ENSMUSG00000053965</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Penk</th>
      <td>ENSMUSG00000045573</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Uchl1</th>
      <td>ENSMUSG00000029223</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Nap1l5</th>
      <td>ENSMUSG00000055430</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Slc17a6</th>
      <td>ENSMUSG00000030500</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Eomes</th>
      <td>ENSMUSG00000032446</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Col6a1</th>
      <td>ENSMUSG00000001119</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Gria1</th>
      <td>ENSMUSG00000020524</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Vtn</th>
      <td>ENSMUSG00000017344</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Tbx21</th>
      <td>ENSMUSG00000001444</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Kctd12</th>
      <td>ENSMUSG00000098557</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Cystm1</th>
      <td>ENSMUSG00000046727</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Camk2a</th>
      <td>ENSMUSG00000024617</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
    <tr>
      <th>Syt7</th>
      <td>ENSMUSG00000024743</td>
      <td>Gene Expression</td>
      <td>mm10</td>
      <td>True</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<p>In this case study, we include the effect of input spatial transcriptome in constructing the cell-cell network.</p>
<ul class="simple">
<li><p>Burn in 5 steps using only histology features (vanilla version) for all highly variable genes.</p></li>
<li><p>Load the burnin result, perform PCA, and use the transcriptome to adjust the cell-cell network.</p></li>
<li><p>The cell-cell network with transcriptome PCA information will be used to perform the production diffusion.</p></li>
</ul>
<p>Run time (~ 3 mins on MacbookPro M2)</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">MOB</span><span class="o">.</span><span class="n">recipe</span> <span class="o">=</span> <span class="s1">&#39;gene&#39;</span>
<span class="n">MOB</span><span class="o">.</span><span class="n">set_params</span><span class="p">(</span>
    <span class="n">is_rawCount</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">out_prefix</span><span class="o">=</span><span class="s2">&quot;fineST&quot;</span><span class="p">,</span>
    <span class="n">write_freq</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span>
    <span class="n">n_iter</span><span class="o">=</span><span class="mi">40</span><span class="p">,</span>
    <span class="n">conn_csr_matrix</span><span class="o">=</span><span class="s2">&quot;force&quot;</span><span class="p">,</span>
    <span class="n">smoothing_scale</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>
    <span class="n">node_features_obs_list</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;spot_heterogeneity&#39;</span><span class="p">],</span>
    <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">geom_morph_ratio</span><span class="o">=</span><span class="mf">0.5</span><span class="p">,</span>
    <span class="n">geom_constraint</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">inflation_percentage</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">regulate_expression_mean</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">stochastic_expression_neighbors_level</span><span class="o">=</span><span class="s1">&#39;spot&#39;</span><span class="p">,</span>
    <span class="n">smooth_predicted_expression_steps</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span>
    <span class="n">reduced_dimension_transcriptome_obsm_key</span><span class="o">=</span><span class="s1">&#39;X_pca&#39;</span><span class="p">,</span>
    <span class="n">adjust_cell_network_by_transcriptome_scale</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>
    <span class="n">n_jobs</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

<span class="n">MOB</span><span class="o">.</span><span class="n">predict_gene_expression</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-cyan-fg">[12/13/23 12:49:30]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Using mode gene
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Creating piles: 100%|<span class="ansi-green-fg">██████████</span>| 4/4 [00:01&lt;00:00,  2.67it/s]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-cyan-fg">[12/13/23 12:52:17]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Forcing to recalculate the connectivities.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Construct SNN with morphological features: <span class="ansi-bold">[</span><span class="ansi-green-fg">&#39;mean_gray&#39;</span>, <span class="ansi-green-fg">&#39;std_gray&#39;</span>, <span class="ansi-green-fg">&#39;entropy_img&#39;</span>, <span class="ansi-green-fg">&#39;mean_r&#39;</span>, <span class="ansi-green-fg">&#39;mean_g&#39;</span>, <span class="ansi-green-fg">&#39;mean_b&#39;</span>,
<span class="ansi-cyan-fg">                    </span>         <span class="ansi-green-fg">&#39;std_r&#39;</span>, <span class="ansi-green-fg">&#39;std_g&#39;</span>, <span class="ansi-green-fg">&#39;std_b&#39;</span><span class="ansi-bold">]</span>.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Incorporate the effect of transcriptome.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Finish constructing SNN
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Add adata.obsp<span class="ansi-bold">[</span><span class="ansi-green-fg">&#34;snn_connectivities&#34;</span><span class="ansi-bold">]</span>
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Add adata.obsp<span class="ansi-bold">[</span><span class="ansi-green-fg">&#34;snn_knn_connectivities&#34;</span><span class="ansi-bold">]</span>
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Add adata.uns<span class="ansi-bold">[</span><span class="ansi-green-fg">&#34;snn&#34;</span><span class="ansi-bold">]</span>
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Weigh cells according to the spot heterogeneity.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Promote flow of information from more homogeneous cells to less.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Eliminate low quality edges <span class="ansi-bold">(</span>&lt;<span class="ansi-cyan-intense-fg ansi-bold">0.1</span><span class="ansi-bold">)</span> between cells.
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Compute transition matrix
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: self weight scale is set to: <span class="ansi-cyan-intense-fg ansi-bold">0.200</span>
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Compute transition matrix
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: self weight scale is set to: <span class="ansi-cyan-intense-fg ansi-bold">1.808</span>
<span class="ansi-cyan-fg">[12/13/23 12:52:18]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: Added transition matrix to adata.obsp<span class="ansi-bold">[</span><span class="ansi-green-fg">&#34;snn_transition_matrix&#34;</span><span class="ansi-bold">]</span>
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
... storing &#39;spot_barcodes&#39; as categorical
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
<span class="ansi-cyan-fg">                   </span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: fineST estimation starts.
<span class="ansi-cyan-fg">[12/13/23 12:52:34]</span><span class="ansi-cyan-fg"> </span><span class="ansi-blue-fg">INFO    </span> Thor: fineST estimation finished.
<span class="ansi-cyan-fg">                    </span>
<span class="ansi-cyan-fg">                    </span>
</pre></div></div>
</div>
<section id="Check-the-expression-of-a-few-genes.">
<h3>Check the expression of a few genes.<a class="headerlink" href="#Check-the-expression-of-a-few-genes." title="Link to this heading"></a></h3>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad_thor</span> <span class="o">=</span> <span class="n">MOB</span><span class="o">.</span><span class="n">load_result</span><span class="p">(</span><span class="s1">&#39;fineST_40_samp10.npz&#39;</span><span class="p">)</span>
<span class="n">ad_thor</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 40543 × 17
    obs: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;, &#39;spot_barcodes&#39;, &#39;x&#39;, &#39;y&#39;, &#39;mean_gray&#39;, &#39;std_gray&#39;, &#39;entropy_img&#39;, &#39;mean_r&#39;, &#39;mean_g&#39;, &#39;mean_b&#39;, &#39;std_r&#39;, &#39;std_g&#39;, &#39;std_b&#39;, &#39;spot_heterogeneity&#39;, &#39;node_weights&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;used_for_prediction&#39;, &#39;used_for_reduced&#39;, &#39;used_for_vae&#39;, &#39;highly_variable&#39;, &#39;means&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;log1p&#39;, &#39;spatial&#39;, &#39;cell_image_props&#39;, &#39;hvg&#39;, &#39;snn&#39;
    obsm: &#39;spatial&#39;, &#39;X_pca&#39;
    obsp: &#39;snn_connectivities&#39;, &#39;snn_knn_connectivities&#39;, &#39;snn_transition_matrix&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Eomes&#39;</span><span class="p">,</span> <span class="s1">&#39;Uchl1&#39;</span><span class="p">,</span> <span class="s1">&#39;Pcp4l1&#39;</span><span class="p">,</span> <span class="s1">&#39;Col6a1&#39;</span><span class="p">,</span> <span class="s1">&#39;Tbx21&#39;</span><span class="p">,</span> <span class="s1">&#39;Slc17a6&#39;</span><span class="p">,</span> <span class="s1">&#39;Bmp7&#39;</span><span class="p">]</span>

<span class="n">ad_spot</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">spot_adata_path</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">ad_spot</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">genes</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="s1">&#39;p99&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_MOB_26_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_MOB_26_0.png" style="width: 1180px; height: 567px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Eomes&#39;</span><span class="p">,</span> <span class="s1">&#39;Uchl1&#39;</span><span class="p">,</span> <span class="s1">&#39;Pcp4l1&#39;</span><span class="p">]</span>

<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">scanpy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">dpi_save</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">fw2</span> <span class="o">=</span> <span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span><span class="s1">&#39;fw2&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">.</span><span class="n">continuous_palettes</span><span class="p">[</span><span class="s1">&#39;fireworks2&#39;</span><span class="p">])</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">15</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>

<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">ad_thor</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene</span><span class="p">,</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="s1">&#39;p5&#39;</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="s1">&#39;p95&#39;</span><span class="p">,</span> <span class="n">colorbar_loc</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="s1">&#39;coolwarm&#39;</span><span class="p">,</span> <span class="n">img_key</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>
    <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_MOB_27_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_MOB_27_0.png" style="width: 881px; height: 338px;" />
</div>
</div>
<p>Visualize module-specific genes</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">matplotlib.colors</span> <span class="kn">import</span> <span class="n">LinearSegmentedColormap</span>
<span class="kn">from</span> <span class="nn">thor.pl</span> <span class="kn">import</span> <span class="n">colors</span>

<span class="n">mod_genes</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Col6a1&#39;</span><span class="p">,</span> <span class="s1">&#39;Tbx21&#39;</span><span class="p">,</span> <span class="s1">&#39;Slc17a6&#39;</span><span class="p">,</span> <span class="s1">&#39;Bmp7&#39;</span><span class="p">]</span>


<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">scanpy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">dpi</span><span class="o">=</span><span class="mi">80</span><span class="p">,</span> <span class="n">dpi_save</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span> <span class="n">facecolor</span><span class="o">=</span><span class="s1">&#39;k&#39;</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="n">yb</span> <span class="o">=</span> <span class="n">LinearSegmentedColormap</span><span class="o">.</span><span class="n">from_list</span><span class="p">(</span><span class="s1">&#39;yb&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">.</span><span class="n">continuous_palettes</span><span class="p">[</span><span class="s1">&#39;blueYellow&#39;</span><span class="p">])</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">20</span><span class="p">,</span> <span class="mi">5</span><span class="p">))</span>
<span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>

<span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">mod_genes</span><span class="p">:</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial</span><span class="p">(</span><span class="n">ad_thor</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">gene</span><span class="p">,</span> <span class="n">spot_size</span><span class="o">=</span><span class="mi">15</span><span class="p">,</span> <span class="n">vmin</span><span class="o">=</span><span class="s1">&#39;p2&#39;</span><span class="p">,</span> <span class="n">vmax</span><span class="o">=</span><span class="s1">&#39;p98&#39;</span><span class="p">,</span> <span class="n">colorbar_loc</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">cmap</span><span class="o">=</span><span class="n">yb</span><span class="p">,</span> <span class="n">img_key</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">])</span>
    <span class="n">axes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;w&#39;</span><span class="p">)</span>
    <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_runThor_MOB_29_0.png" class="no-scaled-link" src="../_images/notebooks_runThor_MOB_29_0.png" style="width: 1197px; height: 338px;" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../_autosummary/thor.pl.fringe.html" class="btn btn-neutral float-left" title="thor.pl.fringe" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="runThor_DCIS.html" class="btn btn-neutral float-right" title="Ductal carcinoma in situ" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Wang Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>