<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Classes &mdash; Thor  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=92fd9be5" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="_static/style.css?v=71a31e8f" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=5929fcd5"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="thor.fineST" href="_autosummary/thor.fineST.html" />
    <link rel="prev" title="Frequently asked questions" href="FAQ.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            Thor
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Thor</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Thor Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="release_notes.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="FAQ.html">Frequently asked questions</a></li>
</ul>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Classes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#prediction-of-in-silico-cell-gene-expression">Prediction of in silico cell gene expression</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.fineST.html">thor.fineST</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#preprocessing-of-wsi">Preprocessing of WSI</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pp.WholeSlideImage.html">thor.pp.WholeSlideImage</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#preprocessing-of-spatial-transcriptomics">Preprocessing of spatial transcriptomics</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pp.Spatial.html">thor.pp.Spatial</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#example-interface-with-external-r-packages">Example interface with external R packages</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.analy.SPARKX.html">thor.analy.SPARKX</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="#api">API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#preprocessing">Preprocessing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pp.preprocess_image.html">thor.pp.preprocess_image</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pp.load_nuclei.html">thor.pp.load_nuclei</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pp.load_cellpose.html">thor.pp.load_cellpose</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pp.load_cellprofiler.html">thor.pp.load_cellprofiler</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pp.load_mask_npz.html">thor.pp.load_mask_npz</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pp.nuclei_segmentation.html">thor.pp.nuclei_segmentation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#advanced-analyses">Advanced analyses</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.analy.analyze_gene_expression_gradient.html">thor.analy.analyze_gene_expression_gradient</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.analy.compute_dge_between_regions.html">thor.analy.compute_dge_between_regions</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.analy.get_pathway_score.html">thor.analy.get_pathway_score</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.analy.get_tf_activity.html">thor.analy.get_tf_activity</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.analy.read_polygon_ROI.html">thor.analy.read_polygon_ROI</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.analy.prepare_and_run_copykat.html">thor.analy.prepare_and_run_copykat</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.analy.adata_to_mtx_conversion.html">thor.analy.adata_to_mtx_conversion</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.analy.run_commot.html">thor.analy.run_commot</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#plotting">Plotting</a><ul>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pl.single.html">thor.pl.single</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pl.multiple.html">thor.pl.multiple</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pl.single_molecule.html">thor.pl.single_molecule</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pl.multi_molecules.html">thor.pl.multi_molecules</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pl.clusters.html">thor.pl.clusters</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pl.spot_over.html">thor.pl.spot_over</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pl.annotate_ROI.html">thor.pl.annotate_ROI</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pl.deg.html">thor.pl.deg</a></li>
<li class="toctree-l3"><a class="reference internal" href="_autosummary/thor.pl.fringe.html">thor.pl.fringe</a></li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Tutorials</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="notebooks/runThor_MOB.html">Mouse olfactory bulb</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/runThor_DCIS.html">Ductal carcinoma in situ</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/run_thor_analyses_HF.html">Human heart failure</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/tutorial_runThor_VisiumHD.html">Run Thor on Visium HD data</a></li>
<li class="toctree-l1"><a class="reference internal" href="notebooks/tutorial_run_cell_communication_commot.html">Cell cell communication</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Mjolnir visualization</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="Mjolnir.html">About Mjolnir</a></li>
<li class="toctree-l1"><a class="reference internal" href="Mjolnir_installation.html">Mjolnir Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Mjolnir_video_tutorials.html">Mjolnir video tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="Mjolnir_test_data.html">Mjolnir test Data</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Thor</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Classes</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/API.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <p>Import Thor as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">thor</span>
</pre></div>
</div>
<section id="classes">
<h1>Classes<a class="headerlink" href="#classes" title="Link to this heading"></a></h1>
<section id="prediction-of-in-silico-cell-gene-expression">
<h2>Prediction of in silico cell gene expression<a class="headerlink" href="#prediction-of-in-silico-cell-gene-expression" title="Link to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.fineST.html#thor.fineST" title="thor.fineST"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.fineST</span></code></a>(image_path, name[, ...])</p></td>
<td><p>Class for in silico cell gene expression inference</p></td>
</tr>
</tbody>
</table>
</section>
<section id="preprocessing-of-wsi">
<h2>Preprocessing of WSI<a class="headerlink" href="#preprocessing-of-wsi" title="Link to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.pp.WholeSlideImage.html#thor.pp.WholeSlideImage" title="thor.pp.WholeSlideImage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pp.WholeSlideImage</span></code></a>(image_path[, name, ...])</p></td>
<td><p>Whole slide image class.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="preprocessing-of-spatial-transcriptomics">
<h2>Preprocessing of spatial transcriptomics<a class="headerlink" href="#preprocessing-of-spatial-transcriptomics" title="Link to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.pp.Spatial.html#thor.pp.Spatial" title="thor.pp.Spatial"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pp.Spatial</span></code></a>(name, st_dir[, image_path, ...])</p></td>
<td><p>Class for spatial transcriptomics data.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="example-interface-with-external-r-packages">
<h2>Example interface with external R packages<a class="headerlink" href="#example-interface-with-external-r-packages" title="Link to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.analy.SPARKX.html#thor.analy.SPARKX" title="thor.analy.SPARKX"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.analy.SPARKX</span></code></a>([rscript_path])</p></td>
<td><p>Class for running <a class="reference external" href="https://doi.org/10.1186/s13059-021-02404-0">SPARK-X</a>.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="api">
<h1>API<a class="headerlink" href="#api" title="Link to this heading"></a></h1>
<section id="preprocessing">
<h2>Preprocessing<a class="headerlink" href="#preprocessing" title="Link to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.pp.preprocess_image.html#thor.pp.preprocess_image" title="thor.pp.preprocess_image"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pp.preprocess_image</span></code></a>(image_path[, bbox, ...])</p></td>
<td><p>Preprocess the image and extract features from the cells.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/thor.pp.load_nuclei.html#thor.pp.load_nuclei" title="thor.pp.load_nuclei"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pp.load_nuclei</span></code></a>([nuclei_path, source_format])</p></td>
<td><p>Load nuclei segmentation result from a file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.pp.load_cellpose.html#thor.pp.load_cellpose" title="thor.pp.load_cellpose"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pp.load_cellpose</span></code></a>(nuclei_path)</p></td>
<td><p>Load nuclei segmentation result from a cellpose output file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/thor.pp.load_cellprofiler.html#thor.pp.load_cellprofiler" title="thor.pp.load_cellprofiler"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pp.load_cellprofiler</span></code></a>(nuclei_path)</p></td>
<td><p>Load nuclei segmentation result from a cellprofiler output file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.pp.load_mask_npz.html#thor.pp.load_mask_npz" title="thor.pp.load_mask_npz"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pp.load_mask_npz</span></code></a>(nuclei_path)</p></td>
<td><p>Load nuclei segmentation result from a mask array npz file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/thor.pp.nuclei_segmentation.html#thor.pp.nuclei_segmentation" title="thor.pp.nuclei_segmentation"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pp.nuclei_segmentation</span></code></a>(image_path[, ...])</p></td>
<td><p>Segment nuclei from H&amp;E stained images using stardist, cellpose or histocartography.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="advanced-analyses">
<h2>Advanced analyses<a class="headerlink" href="#advanced-analyses" title="Link to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.analy.analyze_gene_expression_gradient.html#thor.analy.analyze_gene_expression_gradient" title="thor.analy.analyze_gene_expression_gradient"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.analy.analyze_gene_expression_gradient</span></code></a>(adata)</p></td>
<td><p>Analyze gene expression against a baseline in a selected region of interest (ROI).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/thor.analy.compute_dge_between_regions.html#thor.analy.compute_dge_between_regions" title="thor.analy.compute_dge_between_regions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.analy.compute_dge_between_regions</span></code></a>(...)</p></td>
<td><p>Compute differential gene expression (DGE) between two regions.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.analy.get_pathway_score.html#thor.analy.get_pathway_score" title="thor.analy.get_pathway_score"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.analy.get_pathway_score</span></code></a>(adata[, layer, ...])</p></td>
<td><p>Calculate pathway score for each cell using over-representation analysis.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/thor.analy.get_tf_activity.html#thor.analy.get_tf_activity" title="thor.analy.get_tf_activity"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.analy.get_tf_activity</span></code></a>(adata[, layer, ...])</p></td>
<td><p>Infer TF activity using the CollecTRI database.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.analy.read_polygon_ROI.html#thor.analy.read_polygon_ROI" title="thor.analy.read_polygon_ROI"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.analy.read_polygon_ROI</span></code></a>(json_path, adata)</p></td>
<td><p>Read polygon ROI from json file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/thor.analy.prepare_and_run_copykat.html#thor.analy.prepare_and_run_copykat" title="thor.analy.prepare_and_run_copykat"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.analy.prepare_and_run_copykat</span></code></a>(adata[, ...])</p></td>
<td><p>Run CopyKAT on the input data.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.analy.adata_to_mtx_conversion.html#thor.analy.adata_to_mtx_conversion" title="thor.analy.adata_to_mtx_conversion"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.analy.adata_to_mtx_conversion</span></code></a>(adata[, ...])</p></td>
<td><p>Convert AnnData object to matrix market format.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/thor.analy.run_commot.html#thor.analy.run_commot" title="thor.analy.run_commot"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.analy.run_commot</span></code></a>(adata[, region, ...])</p></td>
<td><p>Run the cell-cell communication analysis using the modified <a class="reference external" href="https://doi.org/10.1038/s41592-022-01728-4">COMMOT</a> method.</p></td>
</tr>
</tbody>
</table>
</section>
<section id="plotting">
<h2>Plotting<a class="headerlink" href="#plotting" title="Link to this heading"></a></h2>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.pl.single.html#thor.pl.single" title="thor.pl.single"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pl.single</span></code></a>(var_name, img_mask[, ax, ...])</p></td>
<td><p>Color the cells or nuclei with one variable, gene name, or any other observable.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/thor.pl.multiple.html#thor.pl.multiple" title="thor.pl.multiple"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pl.multiple</span></code></a>(vars_list, img_masks_list)</p></td>
<td><p>Color the cells or nuclei with multiple variables.</p></td>
</tr>
</tbody>
</table>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.pl.single_molecule.html#thor.pl.single_molecule" title="thor.pl.single_molecule"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pl.single_molecule</span></code></a>(var, var_expression, ...)</p></td>
<td><p>Color the cells or nuclei with a variable, gene or any observable (gene expression vector as input).</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/thor.pl.multi_molecules.html#thor.pl.multi_molecules" title="thor.pl.multi_molecules"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pl.multi_molecules</span></code></a>(vars_list, ...[, ...])</p></td>
<td><p>Color the cells or nuclei with multiple variables, gene or any observable (gene expression array as input).</p></td>
</tr>
</tbody>
</table>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.pl.clusters.html#thor.pl.clusters" title="thor.pl.clusters"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pl.clusters</span></code></a>(cluster_labels_list[, ...])</p></td>
<td><p>Color the cells or nuclei with cluster labels.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/thor.pl.spot_over.html#thor.pl.spot_over" title="thor.pl.spot_over"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pl.spot_over</span></code></a>(ad, ad_spot[, spot_scale, ...])</p></td>
<td><p>Plot spatial expression data with spots on top.</p></td>
</tr>
</tbody>
</table>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.pl.annotate_ROI.html#thor.pl.annotate_ROI" title="thor.pl.annotate_ROI"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pl.annotate_ROI</span></code></a>(im[, ROI_polygon, ...])</p></td>
<td><p>Annotate the ROI and baseline on the image.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="_autosummary/thor.pl.deg.html#thor.pl.deg" title="thor.pl.deg"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pl.deg</span></code></a>([data, genes, ...])</p></td>
<td><p>Plot log2foldchange of gene expression against distance from the baseline.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="_autosummary/thor.pl.fringe.html#thor.pl.fringe" title="thor.pl.fringe"><code class="xref py py-obj docutils literal notranslate"><span class="pre">thor.pl.fringe</span></code></a>([data, genes, cmap, lw, ...])</p></td>
<td><p>Plot log2foldchange of gene expression against distance from the baseline.</p></td>
</tr>
</tbody>
</table>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="FAQ.html" class="btn btn-neutral float-left" title="Frequently asked questions" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="_autosummary/thor.fineST.html" class="btn btn-neutral float-right" title="thor.fineST" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Wang Lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>